%% Generated by Sphinx.
\def\sphinxdocclass{report}
\documentclass[letterpaper,10pt,english]{sphinxmanual}
\ifdefined\pdfpxdimen
   \let\sphinxpxdimen\pdfpxdimen\else\newdimen\sphinxpxdimen
\fi \sphinxpxdimen=.75bp\relax

\usepackage[utf8]{inputenc}
\ifdefined\DeclareUnicodeCharacter
 \ifdefined\DeclareUnicodeCharacterAsOptional
  \DeclareUnicodeCharacter{"00A0}{\nobreakspace}
  \DeclareUnicodeCharacter{"2500}{\sphinxunichar{2500}}
  \DeclareUnicodeCharacter{"2502}{\sphinxunichar{2502}}
  \DeclareUnicodeCharacter{"2514}{\sphinxunichar{2514}}
  \DeclareUnicodeCharacter{"251C}{\sphinxunichar{251C}}
  \DeclareUnicodeCharacter{"2572}{\textbackslash}
 \else
  \DeclareUnicodeCharacter{00A0}{\nobreakspace}
  \DeclareUnicodeCharacter{2500}{\sphinxunichar{2500}}
  \DeclareUnicodeCharacter{2502}{\sphinxunichar{2502}}
  \DeclareUnicodeCharacter{2514}{\sphinxunichar{2514}}
  \DeclareUnicodeCharacter{251C}{\sphinxunichar{251C}}
  \DeclareUnicodeCharacter{2572}{\textbackslash}
 \fi
\fi
\usepackage{cmap}
\usepackage[T1]{fontenc}
\usepackage{amsmath,amssymb,amstext}
\usepackage{babel}
\usepackage{times}
\usepackage[Sonny]{fncychap}
\usepackage{sphinx}

\usepackage{geometry}

% Include hyperref last.
\usepackage{hyperref}
% Fix anchor placement for figures with captions.
\usepackage{hypcap}% it must be loaded after hyperref.
% Set up styles of URL: it should be placed after hyperref.
\urlstyle{same}

\addto\captionsenglish{\renewcommand{\figurename}{Fig.}}
\addto\captionsenglish{\renewcommand{\tablename}{Table}}
\addto\captionsenglish{\renewcommand{\literalblockname}{Listing}}

\addto\captionsenglish{\renewcommand{\literalblockcontinuedname}{continued from previous page}}
\addto\captionsenglish{\renewcommand{\literalblockcontinuesname}{continues on next page}}

\addto\extrasenglish{\def\pageautorefname{page}}

\setcounter{tocdepth}{3}
\setcounter{secnumdepth}{3}


\title{MATCHBox API Utils Documentation}
\date{Feb 27, 2018}
\release{1.0.0}
\author{Dave Sims}
\newcommand{\sphinxlogo}{\vbox{}}
\renewcommand{\releasename}{Release}
\makeindex

\begin{document}

\maketitle
\sphinxtableofcontents
\phantomsection\label{\detokenize{index::doc}}



\chapter{MATCHBox API Utils}
\label{\detokenize{intro::doc}}\label{\detokenize{intro:welcome-to-matchbox-api-utils-documentation}}\label{\detokenize{intro:matchbox-api-utils}}

\section{Introduction}
\label{\detokenize{intro:introduction}}
Retrieving results from the MATCHBox, an NCI derived system to collect, collate,
and report data from the NCI-MATCH study can be a difficult task.  While there is
a rich API and set of tools underlying the user interfact (UI), it can be very
difficult at times to quickly gleen the information one is looking for from clicking
through all of the web pages.  This is where MATCHBox API Utils comes in.

The system is designed to directly pull data from a live MATCHBox instance, and
build a simple JSON file, from which a whole host of methods have been created
to access, filter, interrogate, report, etc. the data.  For example, if you wanted
to know which MSNs were associated with patient identifier (PSN) 13070, you could
click through the GUI, and find the correct data.  Or, with this tool, you can
simply run:

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{k+kn}{from} \PYG{n+nn}{matchbox\PYGZus{}api\PYGZus{}utils} \PYG{k}{import} \PYG{n}{MatchData}
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{data} \PYG{o}{=} \PYG{n}{MatchData}\PYG{p}{(}\PYG{p}{)}
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{data}\PYG{o}{.}\PYG{n}{get\PYGZus{}msn}\PYG{p}{(}\PYG{n}{psn}\PYG{o}{=}\PYG{l+m+mi}{13070}\PYG{p}{)}
\PYG{g+go}{[u\PYGZsq{}MSN31054\PYGZsq{}, u\PYGZsq{}MSN59774\PYGZsq{}]}
\end{sphinxVerbatim}

More on the specifics of how to use these modules and methods a little later.

Also, along with the set of Python modules and methods, included is a set of
\sphinxtitleref{helper} scripts that will act as nice wrappers around some of these basic
methods.  For example, running the \sphinxcode{\sphinxupquote{map\_msn\_psn.py}} script that’s a part of this
package will generate really nice tables of data, allow for batch processing from
either a list input on the commandline or a file containing a set of identifiers
of interest.

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYGZdl{} map\PYGZus{}msn\PYGZus{}psn.py \PYGZhy{}t MSN MSN31054,MSN12724,MSN37895,MSN12104

Getting MSN / PSN mapping data...

PSN,BSN,MSN
PSN10818,T\PYGZhy{}16\PYGZhy{}000029,MSN12104
PSN10955,T\PYGZhy{}16\PYGZhy{}000213,MSN12724
PSN13070,T\PYGZhy{}16\PYGZhy{}002251,MSN31054
PSN13948,T\PYGZhy{}16\PYGZhy{}003010,MSN37895
\end{sphinxVerbatim}


\section{Installation}
\label{\detokenize{intro:installation}}
At this time, the most simple installation is to unzip the provided tarball, and
run:

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{python} \PYG{n}{setup}\PYG{o}{.}\PYG{n}{py} \PYG{n}{install}
\end{sphinxVerbatim}

From here, the installer will add the library modules to your standard Python
library location, put the helper scripts into the standard binary direcotry in
your \sphinxcode{\sphinxupquote{\$PATH}}, and run a post installation script that will set up the appropriate
URLs, username, and password for your system and user.  At the end, a new set of
JSON files will be created in \sphinxcode{\sphinxupquote{\$HOME/.mb\_utils/}} which will contain the initial
MATCH dataset for use.

We recommend re-building this dataset on occassion using the
\sphinxcode{\sphinxupquote{matchbox\_json\_dump.py}} program included in the package for up to date data,
especially as the dataset is not yet locked down. Additionally, one can keep old
sets of these JSON databases for use at any time (e.g comparing version data), by
inputting the custom JSON into the MatchData object (see: \sphinxcode{\sphinxupquote{matchbox\_api\_utils}}
API documentation for details).


\section{Requirements}
\label{\detokenize{intro:requirements}}
These modules are run under Python 2.7.  However, all coding (where possible) is
done with Python 3 in mind, and should be easily portable. This is especially the
case with the main library modules and classes.  Your mileage may vary with the
helper scripts somewhat.

In addition to the above, the package requires the installation of:
\begin{itemize}
\item {} 
\sphinxtitleref{requests \textless{}http://docs.python-requests.org/en/master/\textgreater{}}

\end{itemize}

The setup script should take care of this dependency upon installation.


\section{Using MATCHBox API Utils}
\label{\detokenize{intro:using-matchbox-api-utils}}
Once the installation is run, one can either use the pre-built scripts available
in your systems binary directory (see Helper Scripts documentation), or write
your own scripts using any one of the included methods (see the MATCHBox API Utils
API documentation).


\chapter{MATCHBox API Utils Tutorial}
\label{\detokenize{tutorial:matchbox-api-utils-tutorial}}\label{\detokenize{tutorial::doc}}
Working with MATCHBox API Utils is very easy.  It is simply a matter of loading one
(or all) of the available modules:
\begin{enumerate}
\item {} 
Matchbox

\item {} 
MatchData

\item {} 
TreatmentArms

\end{enumerate}

Most often one will be using the \sphinxcode{\sphinxupquote{MatchData()}} module, as that deals directly
with the majority of the MATCH data that needs to be dealt with.  It directly pulls
data from \sphinxcode{\sphinxupquote{Matchbox}} and \sphinxcode{\sphinxupquote{TreatmentArms}} where need be, and it allows for the
most flexible parsing and reporting of data.  That’s not to say that the other
modules are not useful.  Far from it!  As we’ll see a little later down, there are
a number of useful things that one can do with those as well.

By default modules will load the most recent dataset available in
\sphinxcode{\sphinxupquote{\$HOME/.mb\_utils}} (see API documentation for specifics), and be ready for use.
However, if one wanted to get a live dataset, or one wanted to get data from an
older snapshot of the database, that is all possible too.


\section{Module: Matchbox()}
\label{\detokenize{tutorial:module-matchbox}}
The \sphinxcode{\sphinxupquote{Matchbox}} module is a basic data connector to the live instance of the
MATCHBox system.  Apart from a URL to query and credentials, there are not a
whole lot of options.

In general this module is called by MatchData and TreatmentArms to make the
connection to the live system and pull data.  I’ve included some documentation of
this module, though, in the event that one needs to call on it at some point.

For example, on occassion one might want to get a complete raw MATCHBox dataset,
which has not been parse and filtered, and can be accepted into the other modules
as if they were making a live call.  In this case, one could run

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{k+kn}{from} \PYG{n+nn}{matchbox\PYGZus{}api\PYGZus{}utils} \PYG{k}{import} \PYG{n}{Matchbox}
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{Matchbox}\PYG{p}{(}\PYG{n}{make\PYGZus{}raw}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{raw\PYGZus{}mb\PYGZus{}dataset.json}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\end{sphinxVerbatim}

This will create a raw complete JSON dump of MATCHBox.

Without dumping that data to a JSON file, one could assign a variable to the
Matchbox obj, and then pass it around.  However, MatchData is better for an
entry point here.


\section{Module: MatchData()}
\label{\detokenize{tutorial:module-matchdata}}
This is by far the workhorse module of the package.  In its most basic form, one
needs only load the object into a variable, and then run the host of methods
available

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{k+kn}{from} \PYG{n+nn}{matchbox\PYGZus{}api\PYGZus{}utils} \PYG{k}{import} \PYG{n}{MatchData}
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{data} \PYG{o}{=} \PYG{n}{MatchData}\PYG{p}{(}\PYG{p}{)}
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{data}\PYG{o}{.}\PYG{n}{get\PYGZus{}biopsy\PYGZus{}summary}\PYG{p}{(}\PYG{n}{category}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{progression}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{ret\PYGZus{}type}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{counts}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{g+go}{\PYGZob{}\PYGZsq{}progression\PYGZsq{}: 16\PYGZcb{}}
\end{sphinxVerbatim}

This simply works by loading a processed version of MATCHBox, which has been
optimized to be smaller and more efficient than the raw MATCHBox dataset, and
then having a set of methods to extract portions of the data for cohort type
analysis.

As mentioned above, one can either choose to use the \sphinxtitleref{system\_default} JSON option
for \sphinxcode{\sphinxupquote{MatchData}}, in which case the latest version of the MATCHBox JSON dump will
be loaded, or one can load their own custom file (e.g. if one wanted to load an old
version of the data to see how things changed).  Alternatively (and with the same
warning as above), one can make a live query at the cost of a long load time.

There is also the possibility of loading just one or more patient IDs into the API
and limiting data to that cohort.  However, this is usually only meant for
debugging purposes and filtering on a patient identifier is usually best done at
each method call.


\subsection{Toy Example \#1}
\label{\detokenize{tutorial:toy-example-1}}
You would like to know the disease state of the progression biopsies indicated
above.  You don’t have any identifiers or other information to go on.
\begin{description}
\item[{Here’s a way one might work through it ::}] \leavevmode
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{k+kn}{from} \PYG{n+nn}{matchbox\PYGZus{}api\PYGZus{}utils} \PYG{k}{import} \PYG{n}{MatchData}
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{k+kn}{from} \PYG{n+nn}{pprint} \PYG{k}{import} \PYG{n}{pprint} \PYG{k}{as} \PYG{n}{pp}
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{data} \PYG{o}{=} \PYG{n}{MatchData}\PYG{p}{(}\PYG{p}{)}
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{results} \PYG{o}{=} \PYG{p}{\PYGZob{}}\PYG{p}{\PYGZcb{}}  \PYG{c+c1}{\PYGZsh{} Let\PYGZsq{}s create a dict to put the results in.}
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{biopsies} \PYG{o}{=} \PYG{n}{data}\PYG{o}{.}\PYG{n}{get\PYGZus{}biopsy\PYGZus{}summary}\PYG{p}{(}\PYG{n}{category}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{progression}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{ret\PYGZus{}type}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{ids}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{pp}\PYG{p}{(}\PYG{n}{biopsies}\PYG{p}{)}
\PYG{g+go}{\PYGZob{}\PYGZsq{}progression\PYGZsq{}: [u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}002064\PYGZsq{},}
\PYG{g+go}{                 u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}002755\PYGZsq{},}
\PYG{g+go}{                 u\PYGZsq{}T\PYGZhy{}18\PYGZhy{}000071\PYGZsq{},}
\PYG{g+go}{                 u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}002680\PYGZsq{},}
\PYG{g+go}{                 u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}000787\PYGZsq{},}
\PYG{g+go}{                 u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}002621\PYGZsq{},}
\PYG{g+go}{                 u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}002657\PYGZsq{},}
\PYG{g+go}{                 u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}001275\PYGZsq{},}
\PYG{g+go}{                 u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}001175\PYGZsq{},}
\PYG{g+go}{                 u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}002564\PYGZsq{},}
\PYG{g+go}{                 u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}002556\PYGZsq{},}
\PYG{g+go}{                 u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}002730\PYGZsq{},}
\PYG{g+go}{                 u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}002600\PYGZsq{},}
\PYG{g+go}{                 u\PYGZsq{}T\PYGZhy{}18\PYGZhy{}000005\PYGZsq{},}
\PYG{g+go}{                 u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}000333\PYGZsq{},}
\PYG{g+go}{                 u\PYGZsq{}T\PYGZhy{}18\PYGZhy{}000031\PYGZsq{}]\PYGZcb{}}
\end{sphinxVerbatim}

\end{description}

Notice that we changed the return type of the \sphinxcode{\sphinxupquote{get\_biopsy\_summary()}} call to
\sphinxtitleref{ids}, which allows us to get ids rather than counts.  Now that we have those
biopsy IDs, we can get some PSNs , which will be helpful in getting the disease
data ultimately

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{k}{for} \PYG{n}{bsn} \PYG{o+ow}{in} \PYG{n}{biopsies}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{progression}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{:}
\PYG{g+gp}{... }    \PYG{n}{psn} \PYG{o}{=} \PYG{n}{data}\PYG{o}{.}\PYG{n}{get\PYGZus{}psn}\PYG{p}{(}\PYG{n}{bsn}\PYG{o}{=}\PYG{n}{bsn}\PYG{p}{)}
\PYG{g+gp}{... }    \PYG{n}{msn} \PYG{o}{=} \PYG{n}{data}\PYG{o}{.}\PYG{n}{get\PYGZus{}msn}\PYG{p}{(}\PYG{n}{bsn}\PYG{o}{=}\PYG{n}{bsn}\PYG{p}{)}
\PYG{g+gp}{... }    \PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+si}{\PYGZpc{}s}\PYG{l+s+s1}{: }\PYG{l+s+si}{\PYGZpc{}s}\PYG{l+s+s1}{\PYGZsq{}} \PYG{o}{\PYGZpc{}} \PYG{p}{(}\PYG{n}{psn}\PYG{p}{,} \PYG{n}{msn}\PYG{p}{)}\PYG{p}{)}
\PYG{g+go}{PSN13070: [u\PYGZsq{}MSN59774\PYGZsq{}]}
\PYG{g+go}{PSN13670: [u\PYGZsq{}MSN62646\PYGZsq{}]}
\PYG{g+go}{PSN15436: None}
\PYG{g+go}{PSN15362: [u\PYGZsq{}MSN62489\PYGZsq{}]}
\PYG{g+go}{PSN10955: [u\PYGZsq{}MSN46367\PYGZsq{}]}
\PYG{g+go}{PSN13948: [u\PYGZsq{}MSN62208\PYGZsq{}]}
\PYG{g+go}{PSN11347: [u\PYGZsq{}MSN62398\PYGZsq{}]}
\PYG{g+go}{PSN10818: [u\PYGZsq{}MSN51268\PYGZsq{}]}
\PYG{g+go}{PSN11083: [u\PYGZsq{}MSN50799\PYGZsq{}]}
\PYG{g+go}{PSN11707: [u\PYGZsq{}MSN62042\PYGZsq{}]}
\PYG{g+go}{PSN11769: [u\PYGZsq{}MSN62018\PYGZsq{}]}
\PYG{g+go}{PSN11127: [u\PYGZsq{}MSN62547\PYGZsq{}]}
\PYG{g+go}{PSN12471: [u\PYGZsq{}MSN62109\PYGZsq{}]}
\PYG{g+go}{PSN14705: [u\PYGZsq{}MSN62687\PYGZsq{}]}
\PYG{g+go}{PSN11583: [u\PYGZsq{}MSN41897\PYGZsq{}]}
\PYG{g+go}{PSN15971: None}
\end{sphinxVerbatim}

Looks like there are a couple issues here.
\begin{enumerate}
\item {} 
First, results from the \sphinxcode{\sphinxupquote{get\_msn()}} method are always lists.  We can have
multiple MSNs per BSN unfortunately, and so we need to output more than one
on occassion.  In this case, for what we want to do, we can just output them
all as a comma separated list.

\item {} 
Second some results to not have a MSN returned!  This can happen.  In this
case, there was a progression biopsy collected, but no valid MSN was yet
generated for the case.  Since we would prefer to only work with complete
data for now, we’ll skip those cases.

\end{enumerate}

So, now that we know which are progression cases from the whole dataset, and know
their PSN, BSN, and MSN identifiers, let’s get the disease for each, and store it
in our \sphinxcode{\sphinxupquote{results}} dict above.  We’ll rewrite a little bit of the code above to
help with some of the processing

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{k}{for} \PYG{n}{bsn} \PYG{o+ow}{in} \PYG{n}{biopsies}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{progression}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{:}
\PYG{g+gp}{... }    \PYG{n}{psn} \PYG{o}{=} \PYG{n}{data}\PYG{o}{.}\PYG{n}{get\PYGZus{}psn}\PYG{p}{(}\PYG{n}{bsn}\PYG{o}{=}\PYG{n}{bsn}\PYG{p}{)}
\PYG{g+gp}{... }    \PYG{n}{msn} \PYG{o}{=} \PYG{n}{data}\PYG{o}{.}\PYG{n}{get\PYGZus{}msn}\PYG{p}{(}\PYG{n}{bsn}\PYG{o}{=}\PYG{n}{bsn}\PYG{p}{)}
\PYG{g+gp}{... }    \PYG{k}{if} \PYG{n}{msn} \PYG{o+ow}{is} \PYG{o+ow}{not} \PYG{k+kc}{None}\PYG{p}{:}
\PYG{g+gp}{... }        \PYG{n}{results}\PYG{p}{[}\PYG{n}{psn}\PYG{p}{]} \PYG{o}{=} \PYG{p}{[}\PYG{n}{bsn}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{,}\PYG{l+s+s1}{\PYGZsq{}}\PYG{o}{.}\PYG{n}{join}\PYG{p}{(}\PYG{n}{msn}\PYG{p}{)}\PYG{p}{]}
\PYG{g+gp}{...}
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{pp}\PYG{p}{(}\PYG{n}{results}\PYG{p}{)}
\PYG{g+go}{\PYGZob{}\PYGZsq{}PSN10818\PYGZsq{}: [u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}001275\PYGZsq{}, u\PYGZsq{}MSN51268\PYGZsq{}],}
\PYG{g+go}{ \PYGZsq{}PSN10955\PYGZsq{}: [u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}000787\PYGZsq{}, u\PYGZsq{}MSN46367\PYGZsq{}],}
\PYG{g+go}{ \PYGZsq{}PSN11083\PYGZsq{}: [u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}001175\PYGZsq{}, u\PYGZsq{}MSN50799\PYGZsq{}],}
\PYG{g+go}{ \PYGZsq{}PSN11127\PYGZsq{}: [u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}002730\PYGZsq{}, u\PYGZsq{}MSN62547\PYGZsq{}],}
\PYG{g+go}{ \PYGZsq{}PSN11347\PYGZsq{}: [u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}002657\PYGZsq{}, u\PYGZsq{}MSN62398\PYGZsq{}],}
\PYG{g+go}{ \PYGZsq{}PSN11583\PYGZsq{}: [u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}000333\PYGZsq{}, u\PYGZsq{}MSN41897\PYGZsq{}],}
\PYG{g+go}{ \PYGZsq{}PSN11707\PYGZsq{}: [u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}002564\PYGZsq{}, u\PYGZsq{}MSN62042\PYGZsq{}],}
\PYG{g+go}{ \PYGZsq{}PSN11769\PYGZsq{}: [u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}002556\PYGZsq{}, u\PYGZsq{}MSN62018\PYGZsq{}],}
\PYG{g+go}{ \PYGZsq{}PSN12471\PYGZsq{}: [u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}002600\PYGZsq{}, u\PYGZsq{}MSN62109\PYGZsq{}],}
\PYG{g+go}{ \PYGZsq{}PSN13070\PYGZsq{}: [u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}002064\PYGZsq{}, u\PYGZsq{}MSN59774\PYGZsq{}],}
\PYG{g+go}{ \PYGZsq{}PSN13670\PYGZsq{}: [u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}002755\PYGZsq{}, u\PYGZsq{}MSN62646\PYGZsq{}],}
\PYG{g+go}{ \PYGZsq{}PSN13948\PYGZsq{}: [u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}002621\PYGZsq{}, u\PYGZsq{}MSN62208\PYGZsq{}],}
\PYG{g+go}{ \PYGZsq{}PSN14705\PYGZsq{}: [u\PYGZsq{}T\PYGZhy{}18\PYGZhy{}000005\PYGZsq{}, u\PYGZsq{}MSN62687\PYGZsq{}],}
\PYG{g+go}{ \PYGZsq{}PSN15362\PYGZsq{}: [u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}002680\PYGZsq{}, u\PYGZsq{}MSN62489\PYGZsq{}]\PYGZcb{}}
\end{sphinxVerbatim}

Much better!  Now, let’s leverage another method \sphinxcode{\sphinxupquote{get\_hisology()}} to get the
patient’s disease and add it to the data

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{k}{for} \PYG{n}{p} \PYG{o+ow}{in} \PYG{n}{results}\PYG{p}{:}
\PYG{g+gp}{... }    \PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{data}\PYG{o}{.}\PYG{n}{get\PYGZus{}histology}\PYG{p}{(}\PYG{n}{psn}\PYG{o}{=}\PYG{n}{p}\PYG{p}{)}\PYG{p}{)}
\PYG{g+go}{\PYGZob{}\PYGZsq{}PSN15362\PYGZsq{}: u\PYGZsq{}Salivary gland cancer\PYGZsq{}\PYGZcb{}}
\PYG{g+go}{\PYGZob{}\PYGZsq{}PSN11583\PYGZsq{}: u\PYGZsq{}Salivary gland cancer\PYGZsq{}\PYGZcb{}}
\PYG{g+go}{\PYGZob{}\PYGZsq{}PSN13070\PYGZsq{}: u\PYGZsq{}Cholangiocarcinoma, intrahepatic and extrahepatic bile ducts (adenocarcinoma)\PYGZsq{}\PYGZcb{}}
\PYG{g+go}{\PYGZob{}\PYGZsq{}PSN10955\PYGZsq{}: u\PYGZsq{}Squamous cell carcinoma of the anus\PYGZsq{}\PYGZcb{}}
\PYG{g+go}{\PYGZob{}\PYGZsq{}PSN12471\PYGZsq{}: u\PYGZsq{}Carcinosarcoma of the uterus\PYGZsq{}\PYGZcb{}}
\PYG{g+go}{\PYGZob{}\PYGZsq{}PSN10818\PYGZsq{}: u\PYGZsq{}Colorectal cancer, NOS\PYGZsq{}\PYGZcb{}}
\PYG{g+go}{\PYGZob{}\PYGZsq{}PSN11769\PYGZsq{}: u\PYGZsq{}Renal cell carcinoma, clear cell adenocarcinoma\PYGZsq{}\PYGZcb{}}
\PYG{g+go}{\PYGZob{}\PYGZsq{}PSN11347\PYGZsq{}: u\PYGZsq{}Salivary gland cancer\PYGZsq{}\PYGZcb{}}
\PYG{g+go}{\PYGZob{}\PYGZsq{}PSN13948\PYGZsq{}: u\PYGZsq{}Adenocarcinoma of the rectum\PYGZsq{}\PYGZcb{}}
\PYG{g+go}{\PYGZob{}\PYGZsq{}PSN13670\PYGZsq{}: u\PYGZsq{}Ovarian epithelial cancer\PYGZsq{}\PYGZcb{}}
\PYG{g+go}{\PYGZob{}\PYGZsq{}PSN11707\PYGZsq{}: u\PYGZsq{}Cholangiocarcinoma, intrahepatic and extrahepatic bile ducts (adenocarcinoma)\PYGZsq{}\PYGZcb{}}
\PYG{g+go}{\PYGZob{}\PYGZsq{}PSN11083\PYGZsq{}: u\PYGZsq{}Adenocarcinoma of the colon\PYGZsq{}\PYGZcb{}}
\PYG{g+go}{\PYGZob{}\PYGZsq{}PSN14705\PYGZsq{}: u\PYGZsq{}Laryngeal squamous cell carcinoma\PYGZsq{}\PYGZcb{}}
\PYG{g+go}{\PYGZob{}\PYGZsq{}PSN11127\PYGZsq{}: u\PYGZsq{}Invasive breast carcinoma\PYGZsq{}\PYGZcb{}}
\end{sphinxVerbatim}

As we can see the results for this method call are all dicts of \sphinxtitleref{PSN : Disease}
mappings.  So, we can use the PSN to pull the disease and add it to the results

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{k}{for} \PYG{n}{p} \PYG{o+ow}{in} \PYG{n}{results}\PYG{p}{:}
\PYG{g+gp}{... }    \PYG{n}{results}\PYG{p}{[}\PYG{n}{p}\PYG{p}{]}\PYG{o}{.}\PYG{n}{append}\PYG{p}{(}\PYG{n}{data}\PYG{o}{.}\PYG{n}{get\PYGZus{}histology}\PYG{p}{(}\PYG{n}{psn}\PYG{o}{=}\PYG{n}{p}\PYG{p}{)}\PYG{p}{[}\PYG{n}{p}\PYG{p}{]}\PYG{p}{)}
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{pp}\PYG{p}{(}\PYG{n}{results}\PYG{p}{)}
\PYG{g+go}{\PYGZob{}\PYGZsq{}PSN10818\PYGZsq{}: [u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}001275\PYGZsq{}, u\PYGZsq{}MSN51268\PYGZsq{}, u\PYGZsq{}Colorectal cancer, NOS\PYGZsq{}],}
\PYG{g+go}{ \PYGZsq{}PSN10955\PYGZsq{}: [u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}000787\PYGZsq{},}
\PYG{g+go}{              u\PYGZsq{}MSN46367\PYGZsq{},}
\PYG{g+go}{              u\PYGZsq{}Squamous cell carcinoma of the anus\PYGZsq{}],}
\PYG{g+go}{\PYGZsq{}PSN11083\PYGZsq{}: [u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}001175\PYGZsq{}, u\PYGZsq{}MSN50799\PYGZsq{}, u\PYGZsq{}Adenocarcinoma of the colon\PYGZsq{}],}
\PYG{g+go}{\PYGZsq{}PSN11127\PYGZsq{}: [u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}002730\PYGZsq{}, u\PYGZsq{}MSN62547\PYGZsq{}, u\PYGZsq{}Invasive breast carcinoma\PYGZsq{}],}
\PYG{g+go}{\PYGZsq{}PSN11347\PYGZsq{}: [u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}002657\PYGZsq{}, u\PYGZsq{}MSN62398\PYGZsq{}, u\PYGZsq{}Salivary gland cancer\PYGZsq{}],}
\PYG{g+go}{\PYGZsq{}PSN11583\PYGZsq{}: [u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}000333\PYGZsq{}, u\PYGZsq{}MSN41897\PYGZsq{}, u\PYGZsq{}Salivary gland cancer\PYGZsq{}],}
\PYG{g+go}{\PYGZsq{}PSN11707\PYGZsq{}: [u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}002564\PYGZsq{},}
\PYG{g+go}{             u\PYGZsq{}MSN62042\PYGZsq{},}
\PYG{g+go}{             u\PYGZsq{}Cholangiocarcinoma, intrahepatic and extrahepatic bile ducts (adenocarcinoma)\PYGZsq{}],}
\PYG{g+go}{\PYGZsq{}PSN11769\PYGZsq{}: [u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}002556\PYGZsq{},}
\PYG{g+go}{             u\PYGZsq{}MSN62018\PYGZsq{},}
\PYG{g+go}{             u\PYGZsq{}Renal cell carcinoma, clear cell adenocarcinoma\PYGZsq{}],}
\PYG{g+go}{\PYGZsq{}PSN12471\PYGZsq{}: [u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}002600\PYGZsq{}, u\PYGZsq{}MSN62109\PYGZsq{}, u\PYGZsq{}Carcinosarcoma of the uterus\PYGZsq{}],}
\PYG{g+go}{\PYGZsq{}PSN13070\PYGZsq{}: [u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}002064\PYGZsq{},}
\PYG{g+go}{             u\PYGZsq{}MSN59774\PYGZsq{},}
\PYG{g+go}{             u\PYGZsq{}Cholangiocarcinoma, intrahepatic and extrahepatic bile ducts (adenocarcinoma)\PYGZsq{}],}
\PYG{g+go}{\PYGZsq{}PSN13670\PYGZsq{}: [u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}002755\PYGZsq{}, u\PYGZsq{}MSN62646\PYGZsq{}, u\PYGZsq{}Ovarian epithelial cancer\PYGZsq{}],}
\PYG{g+go}{\PYGZsq{}PSN13948\PYGZsq{}: [u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}002621\PYGZsq{}, u\PYGZsq{}MSN62208\PYGZsq{}, u\PYGZsq{}Adenocarcinoma of the rectum\PYGZsq{}],}
\PYG{g+go}{\PYGZsq{}PSN14705\PYGZsq{}: [u\PYGZsq{}T\PYGZhy{}18\PYGZhy{}000005\PYGZsq{},}
\PYG{g+go}{             u\PYGZsq{}MSN62687\PYGZsq{},}
\PYG{g+go}{             u\PYGZsq{}Laryngeal squamous cell carcinoma\PYGZsq{}],}
\PYG{g+go}{\PYGZsq{}PSN15362\PYGZsq{}: [u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}002680\PYGZsq{}, u\PYGZsq{}MSN62489\PYGZsq{}, u\PYGZsq{}Salivary gland cancer\PYGZsq{}]\PYGZcb{}}
\end{sphinxVerbatim}

And finally we have a nice list of collected data for each progression case, which
is ready to print out for downstream use

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{k}{for} \PYG{n}{patient} \PYG{o+ow}{in} \PYG{n}{results}\PYG{p}{:}
\PYG{g+gp}{... }    \PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+se}{\PYGZbs{}t}\PYG{l+s+s1}{\PYGZsq{}}\PYG{o}{.}\PYG{n}{join}\PYG{p}{(}\PYG{p}{[}\PYG{n}{patient}\PYG{p}{]} \PYG{o}{+} \PYG{n}{results}\PYG{p}{[}\PYG{n}{patient}\PYG{p}{]}\PYG{p}{)}\PYG{p}{)}
\PYG{g+go}{PSN15362    T\PYGZhy{}17\PYGZhy{}002680    MSN62489    Salivary gland cancer}
\PYG{g+go}{PSN11583    T\PYGZhy{}17\PYGZhy{}000333    MSN41897    Salivary gland cancer}
\PYG{g+go}{PSN13070    T\PYGZhy{}17\PYGZhy{}002064    MSN59774    Cholangiocarcinoma, intrahepatic and extrahepatic bile ducts (adenocarcinoma)}
\PYG{g+go}{PSN10955    T\PYGZhy{}17\PYGZhy{}000787    MSN46367    Squamous cell carcinoma of the anus}
\PYG{g+go}{PSN12471    T\PYGZhy{}17\PYGZhy{}002600    MSN62109    Carcinosarcoma of the uterus}
\PYG{g+go}{PSN10818    T\PYGZhy{}17\PYGZhy{}001275    MSN51268    Colorectal cancer, NOS}
\PYG{g+go}{PSN11769    T\PYGZhy{}17\PYGZhy{}002556    MSN62018    Renal cell carcinoma, clear cell adenocarcinoma}
\PYG{g+go}{PSN11347    T\PYGZhy{}17\PYGZhy{}002657    MSN62398    Salivary gland cancer}
\PYG{g+go}{PSN13948    T\PYGZhy{}17\PYGZhy{}002621    MSN62208    Adenocarcinoma of the rectum}
\PYG{g+go}{PSN13670    T\PYGZhy{}17\PYGZhy{}002755    MSN62646    Ovarian epithelial cancer}
\PYG{g+go}{PSN11707    T\PYGZhy{}17\PYGZhy{}002564    MSN62042    Cholangiocarcinoma, intrahepatic and extrahepatic bile ducts (adenocarcinoma)}
\PYG{g+go}{PSN11083    T\PYGZhy{}17\PYGZhy{}001175    MSN50799    Adenocarcinoma of the colon}
\PYG{g+go}{PSN14705    T\PYGZhy{}18\PYGZhy{}000005    MSN62687    Laryngeal squamous cell carcinoma}
\PYG{g+go}{PSN11127    T\PYGZhy{}17\PYGZhy{}002730    MSN62547    Invasive breast carcinoma}
\end{sphinxVerbatim}

So, there you have it.  Very simple toy case, but hopefully one that highlights
some of the features of the MatchData module.  See the API documentation section
for more information about the included modules and their usage.


\section{Module TreatmentArms()}
\label{\detokenize{tutorial:module-treatmentarms}}
\textless{}\textless{}Documentation to be written\textgreater{}\textgreater{}


\chapter{MATCHBox API Utils Helper Scripts}
\label{\detokenize{helper_scripts:matchbox-api-utils-helper-scripts}}\label{\detokenize{helper_scripts::doc}}

\section{MATCHBox JSON Dump (matchbox\_json\_dump.py)}
\label{\detokenize{helper_scripts:matchbox-json-dump-matchbox-json-dump-py}}
Get parsed dataset from MATCHBox and dump as a JSON object that we can use
later on to speed up development and periodic searching for data.

This is the script that is run during the post installer, and something that
should be run on a periodic basis to take advantage of new MATCHBox data entered
into the system (at least until lockdown occurs).  This will put the resultant
JSON file into the current directory, where it should probably be migrated into
\sphinxcode{\sphinxupquote{\$HOME/.mb\_utils/}} to be picked up by the rest of the system.


\subsection{MATCHBox JSON Dump Help Doc}
\label{\detokenize{helper_scripts:matchbox-json-dump-help-doc}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{Usage}\PYG{p}{:}
    \PYG{n}{matchbox\PYGZus{}json\PYGZus{}dump}\PYG{o}{.}\PYG{n}{py} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{h}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{d} \PYG{o}{\PYGZlt{}}\PYG{n}{raw\PYGZus{}mb\PYGZus{}datafile}\PYG{o}{.}\PYG{n}{json}\PYG{o}{\PYGZgt{}}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{r}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{p} \PYG{o}{\PYGZlt{}}\PYG{n}{psn}\PYG{o}{\PYGZgt{}}\PYG{p}{]}
    \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{t} \PYG{o}{\PYGZlt{}}\PYG{n}{ta\PYGZus{}obj}\PYG{o}{.}\PYG{n}{json}\PYG{o}{\PYGZgt{}}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{a} \PYG{o}{\PYGZlt{}}\PYG{n}{amoi\PYGZus{}obj}\PYG{o}{.}\PYG{n}{json}\PYG{o}{\PYGZgt{}}\PYG{p}{]}
    \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{m} \PYG{o}{\PYGZlt{}}\PYG{n}{mb\PYGZus{}obj}\PYG{o}{.}\PYG{n}{json}\PYG{o}{\PYGZgt{}}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{v}\PYG{p}{]}

\PYG{n}{Optional} \PYG{n}{Arguments}\PYG{p}{:}
    \PYG{o}{\PYGZhy{}}\PYG{n}{r}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{raw}        \PYG{n}{Generate} \PYG{n}{a} \PYG{n}{raw} \PYG{n}{dump} \PYG{n}{of} \PYG{n}{MATCHBox} \PYG{k}{for} \PYG{n}{debugging} \PYG{o+ow}{and} \PYG{n}{dev}
                     \PYG{n}{purposes}\PYG{o}{.}
                     \PYG{n}{No} \PYG{n}{filtering} \PYG{o+ow}{or} \PYG{n}{condensing} \PYG{n}{of} \PYG{n}{data} \PYG{n}{at} \PYG{n+nb}{all}\PYG{o}{.}
    \PYG{o}{\PYGZhy{}}\PYG{n}{d}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{data}       \PYG{n}{Load} \PYG{n}{a} \PYG{n}{raw} \PYG{n}{MATCHBox} \PYG{n}{database} \PYG{n+nb}{file} \PYG{p}{(}\PYG{n}{usually} \PYG{n}{after} \PYG{n}{running}
                     \PYG{k}{with} \PYG{n}{the} \PYG{o}{\PYGZhy{}}\PYG{n}{r} \PYG{n}{option}\PYG{o}{.}
    \PYG{o}{\PYGZhy{}}\PYG{n}{p}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{patient}    \PYG{n}{Patient} \PYG{n}{sequence} \PYG{n}{number} \PYG{n}{used} \PYG{n}{to} \PYG{n}{limit} \PYG{n}{output} \PYG{k}{for} \PYG{n}{testing}
                     \PYG{o+ow}{and} \PYG{n}{dev} \PYG{n}{purposes}
    \PYG{o}{\PYGZhy{}}\PYG{n}{t}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{ta\PYGZus{}json}    \PYG{n}{Treatment} \PYG{n}{Arms} \PYG{n}{obj} \PYG{n}{JSON} \PYG{n}{filename}\PYG{o}{.} \PYG{n}{DEFAULT}\PYG{p}{:}
                     \PYG{n}{ta\PYGZus{}obj\PYGZus{}}\PYG{o}{\PYGZlt{}}\PYG{n}{datestring}\PYG{o}{\PYGZgt{}}\PYG{o}{.}\PYG{n}{json}
    \PYG{o}{\PYGZhy{}}\PYG{n}{a}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{amoi\PYGZus{}json}  \PYG{n}{aMOIs} \PYG{n}{lookup} \PYG{n}{filename}\PYG{o}{.} \PYG{n}{DEFAULT}\PYG{p}{:}
                     \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{amois\PYGZus{}lookup\PYGZus{}\PYGZlt{}datestring\PYGZgt{}.json}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}
    \PYG{o}{\PYGZhy{}}\PYG{n}{m}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{mb\PYGZus{}json}    \PYG{n}{Name} \PYG{n}{of} \PYG{n}{Match} \PYG{n}{Data} \PYG{n}{obj} \PYG{n}{JSON} \PYG{n+nb}{file}\PYG{o}{.} \PYG{n}{DEFAULT}\PYG{p}{:}
                     \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{mb\PYGZus{}obj\PYGZus{}\PYGZlt{}datestring\PYGZgt{}.json}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}

\PYG{o}{\PYGZhy{}}\PYG{n}{h}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{help}           \PYG{n}{Show} \PYG{n}{this} \PYG{n}{help} \PYG{n}{message} \PYG{o+ow}{and} \PYG{n+nb}{exit}
\PYG{o}{\PYGZhy{}}\PYG{n}{v}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{version}        \PYG{n}{Show} \PYG{n}{program}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{s version number and exit}
\end{sphinxVerbatim}


\section{MAP MSN PSN (map\_msn\_psn.py)}
\label{\detokenize{helper_scripts:map-msn-psn-map-msn-psn-py}}
Input a MSN, BSN, or PSN, and return the other identifiers. Useful when trying
to retrieve the correct dataset and you only know one piece of information.

\begin{sphinxadmonition}{note}{Note:}
We are only working with internal BSN, MSN, and PSN numbers for now and
can not return Outside Assay identifiers at this time.
\end{sphinxadmonition}


\subsection{MAP MSN PSN Help Doc}
\label{\detokenize{helper_scripts:map-msn-psn-help-doc}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{Usage}\PYG{p}{:}
    \PYG{n}{map\PYGZus{}msn\PYGZus{}psn}\PYG{o}{.}\PYG{n}{py} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{h}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{j} \PYG{o}{\PYGZlt{}}\PYG{n}{mb\PYGZus{}json\PYGZus{}file}\PYG{o}{\PYGZgt{}}\PYG{p}{]} \PYG{o}{\PYGZhy{}}\PYG{n}{t} \PYG{p}{\PYGZob{}}\PYG{n}{psn}\PYG{p}{,}\PYG{n}{msn}\PYG{p}{,}\PYG{n}{bsn}\PYG{p}{\PYGZcb{}} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{f} \PYG{o}{\PYGZlt{}}\PYG{n}{input\PYGZus{}file}\PYG{o}{\PYGZgt{}}\PYG{p}{]}
    \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{o} \PYG{o}{\PYGZlt{}}\PYG{n}{outfile}\PYG{o}{\PYGZgt{}}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{v}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZlt{}}\PYG{n}{IDs}\PYG{o}{\PYGZgt{}}\PYG{p}{]}

\PYG{n}{Positional} \PYG{n}{Arguments}\PYG{p}{:}
\PYG{n}{IDs}    \PYG{n}{MATCH} \PYG{n}{IDs} \PYG{n}{to} \PYG{n}{query}\PYG{o}{.} \PYG{n}{Can} \PYG{n}{be} \PYG{n}{single} \PYG{o+ow}{or} \PYG{n}{comma} \PYG{n}{separated} \PYG{n+nb}{list}\PYG{o}{.} \PYG{n}{Must} \PYG{n}{be} \PYG{n}{used}
       \PYG{k}{with} \PYG{n}{PSN} \PYG{o+ow}{or} \PYG{n}{MSN} \PYG{n}{option}\PYG{o}{.}

\PYG{n}{Optional} \PYG{n}{Arguments}\PYG{p}{:}
\PYG{o}{\PYGZhy{}}\PYG{n}{j}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{json}     \PYG{n}{Load} \PYG{n}{a} \PYG{n}{MATCHBox} \PYG{n}{JSON} \PYG{n+nb}{file} \PYG{n}{derived} \PYG{k+kn}{from} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{matchbox\PYGZus{}json\PYGZus{}dump.py}\PYG{l+s+s2}{\PYGZdq{}}
               \PYG{n}{instead} \PYG{n}{of} \PYG{n}{a} \PYG{n}{live} \PYG{n}{query}\PYG{o}{.} \PYG{n}{By} \PYG{n}{default} \PYG{n}{will} \PYG{n}{load} \PYG{n}{the} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{sys\PYGZus{}default}\PYG{l+s+s2}{\PYGZdq{}}
               \PYG{n}{created} \PYG{n}{during} \PYG{n}{package} \PYG{n}{installation}\PYG{o}{.} \PYG{n}{If} \PYG{n}{you} \PYG{n}{wish} \PYG{n}{to} \PYG{n}{do} \PYG{n}{a} \PYG{n}{live}
               \PYG{n}{query} \PYG{p}{(}\PYG{n}{i}\PYG{o}{.}\PYG{n}{e}\PYG{o}{.} \PYG{o+ow}{not} \PYG{n}{load} \PYG{n}{a} \PYG{n}{previously} \PYG{n}{downloaded} \PYG{n}{JSON} \PYG{n}{dump}\PYG{p}{)}\PYG{p}{,} \PYG{n+nb}{set}
               \PYG{o}{\PYGZhy{}}\PYG{n}{j} \PYG{n}{to} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{None}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}
\PYG{o}{\PYGZhy{}}\PYG{n}{t}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n+nb}{type}     \PYG{n}{Type} \PYG{n}{of} \PYG{n}{query} \PYG{n}{string} \PYG{n+nb}{input}\PYG{o}{.} \PYG{n}{Can} \PYG{n}{only} \PYG{n}{be} \PYG{n}{MSN}\PYG{p}{,} \PYG{n}{PSN}\PYG{p}{,} \PYG{o+ow}{or} \PYG{n}{BSN}
\PYG{o}{\PYGZhy{}}\PYG{n}{f}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n+nb}{file}     \PYG{n}{Load} \PYG{n}{a} \PYG{n}{batch} \PYG{n+nb}{file} \PYG{n}{of} \PYG{n+nb}{all} \PYG{n}{MSNs} \PYG{o+ow}{or} \PYG{n}{PSNs} \PYG{n}{to} \PYG{n}{proc}
\PYG{o}{\PYGZhy{}}\PYG{n}{o}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{outfile}  \PYG{n}{File} \PYG{n}{to} \PYG{n}{which} \PYG{n}{output} \PYG{n}{should} \PYG{n}{be} \PYG{n}{written}\PYG{o}{.} \PYG{n}{Default}\PYG{p}{:} \PYG{n}{STDOUT}\PYG{o}{.}
\PYG{o}{\PYGZhy{}}\PYG{n}{h}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{help}     \PYG{n}{Show} \PYG{n}{this} \PYG{n}{help} \PYG{n}{message} \PYG{o+ow}{and} \PYG{n+nb}{exit}
\PYG{o}{\PYGZhy{}}\PYG{n}{v}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{version}  \PYG{n}{Show} \PYG{n}{program}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{s version number and exit}
\end{sphinxVerbatim}


\section{MATCH Variant Frequency (match\_variant\_frequency.py)}
\label{\detokenize{helper_scripts:match-variant-frequency-match-variant-frequency-py}}
Input a list of genes by variant type and get back a table of NCI-MATCH hits
that can be further analyzed in Excel or some other tool. Can either input a patient
(or comma separated list of patients) to query, or query the entire dataset. Will
limit the patient set to the non-outside assay results only, as the Outside Assay
data is very unreliable.


\subsection{MATCH Variant Frequency Help Docs}
\label{\detokenize{helper_scripts:match-variant-frequency-help-docs}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{Usage}\PYG{p}{:}
\PYG{n}{match\PYGZus{}variant\PYGZus{}frequency}\PYG{o}{.}\PYG{n}{py} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{h}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{j} \PYG{o}{\PYGZlt{}}\PYG{n}{mb\PYGZus{}json\PYGZus{}file}\PYG{o}{\PYGZgt{}}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{p} \PYG{o}{\PYGZlt{}}\PYG{n}{PSN}\PYG{o}{\PYGZgt{}}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{s} \PYG{o}{\PYGZlt{}}\PYG{n}{gene\PYGZus{}list}\PYG{o}{\PYGZgt{}}\PYG{p}{]}
\PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{c} \PYG{o}{\PYGZlt{}}\PYG{n}{gene\PYGZus{}list}\PYG{o}{\PYGZgt{}}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{f} \PYG{o}{\PYGZlt{}}\PYG{n}{gene\PYGZus{}list}\PYG{o}{\PYGZgt{}}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{i} \PYG{o}{\PYGZlt{}}\PYG{n}{gene\PYGZus{}list}\PYG{o}{\PYGZgt{}}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{style} \PYG{o}{\PYGZlt{}}\PYG{n}{pp}\PYG{p}{,}\PYG{n}{csv}\PYG{p}{,}\PYG{n}{tsv}\PYG{o}{\PYGZgt{}}\PYG{p}{]}
\PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{o} \PYG{o}{\PYGZlt{}}\PYG{n}{output\PYGZus{}file}\PYG{o}{\PYGZgt{}}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{v}\PYG{p}{]}

\PYG{n}{Optional} \PYG{n}{Arguments}\PYG{p}{:}
\PYG{o}{\PYGZhy{}}\PYG{n}{j}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{json}     \PYG{n}{Load} \PYG{n}{a} \PYG{n}{MATCHBox} \PYG{n}{JSON} \PYG{n+nb}{file} \PYG{n}{derived} \PYG{k+kn}{from} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{matchbox\PYGZus{}json\PYGZus{}dump.py}\PYG{l+s+s2}{\PYGZdq{}}
               \PYG{n}{instead} \PYG{n}{of} \PYG{n}{a} \PYG{n}{live} \PYG{n}{query}
\PYG{o}{\PYGZhy{}}\PYG{n}{p}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{psn}      \PYG{n}{Only} \PYG{n}{output} \PYG{n}{data} \PYG{k}{for} \PYG{n}{a} \PYG{n}{specific} \PYG{n}{patient} \PYG{o+ow}{or} \PYG{n}{comma} \PYG{n}{separated} \PYG{n+nb}{list}
               \PYG{n}{of} \PYG{n}{patients}
\PYG{o}{\PYGZhy{}}\PYG{n}{s}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{snv}      \PYG{n}{Comma} \PYG{n}{separated} \PYG{n+nb}{list} \PYG{n}{of} \PYG{n}{SNVs} \PYG{n}{to} \PYG{n}{look} \PYG{n}{up} \PYG{o+ow}{in} \PYG{n}{MATCHBox} \PYG{n}{data}\PYG{o}{.}
\PYG{o}{\PYGZhy{}}\PYG{n}{c}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{cnv}      \PYG{n}{Comma} \PYG{n}{separated} \PYG{n+nb}{list} \PYG{n}{of} \PYG{n}{CNVs} \PYG{n}{to} \PYG{n}{look} \PYG{n}{up} \PYG{o+ow}{in} \PYG{n}{MATCHBox} \PYG{n}{data}\PYG{o}{.}
\PYG{o}{\PYGZhy{}}\PYG{n}{f}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{fusion}   \PYG{n}{Comma} \PYG{n}{separated} \PYG{n+nb}{list} \PYG{n}{of} \PYG{n}{Fusions} \PYG{n}{to} \PYG{n}{look} \PYG{n}{up} \PYG{o+ow}{in} \PYG{n}{MATCHBox} \PYG{n}{data}\PYG{o}{.}
\PYG{o}{\PYGZhy{}}\PYG{n}{i}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{indel}    \PYG{n}{Comma} \PYG{n}{separated} \PYG{n+nb}{list} \PYG{n}{of} \PYG{n}{Fusions} \PYG{n}{to} \PYG{n}{look} \PYG{n}{up} \PYG{o+ow}{in} \PYG{n}{MATCHBox} \PYG{n}{data}\PYG{o}{.}

\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{style}        \PYG{n}{Format} \PYG{k}{for} \PYG{n}{output}\PYG{o}{.} \PYG{n}{Can} \PYG{n}{choose} \PYG{n}{pretty} \PYG{k}{print} \PYG{p}{(}\PYG{n}{pp}\PYG{p}{)}\PYG{p}{,} \PYG{n}{CSV}\PYG{p}{,} \PYG{o+ow}{or} \PYG{n}{TSV}
\PYG{o}{\PYGZhy{}}\PYG{n}{o}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{output}   \PYG{n}{Output} \PYG{n+nb}{file} \PYG{n}{to} \PYG{n}{which} \PYG{n}{to} \PYG{n}{write} \PYG{n}{data}\PYG{o}{.} \PYG{n}{Default} \PYG{o+ow}{is} \PYG{n}{stdout}
\PYG{o}{\PYGZhy{}}\PYG{n}{h}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{help}     \PYG{n}{Show} \PYG{n}{this} \PYG{n}{help} \PYG{n}{message} \PYG{o+ow}{and} \PYG{n+nb}{exit}
\PYG{o}{\PYGZhy{}}\PYG{n}{v}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{version}  \PYG{n}{Show} \PYG{n}{program}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{s version number and exit}
\end{sphinxVerbatim}


\section{MATCHBox Patient Summary (matchbox\_patient\_summary.py)}
\label{\detokenize{helper_scripts:matchbox-patient-summary-matchbox-patient-summary-py}}
Get patient or disease summary statistics and data from the MATCH dataset.


\subsection{MATCHBox Patient Summary Help Docs}
\label{\detokenize{helper_scripts:matchbox-patient-summary-help-docs}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{Usage}\PYG{p}{:}
\PYG{n}{matchbox\PYGZus{}patient\PYGZus{}summary}\PYG{o}{.}\PYG{n}{py} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{h}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{j} \PYG{o}{\PYGZlt{}}\PYG{n}{mb\PYGZus{}json\PYGZus{}file}\PYG{o}{\PYGZgt{}}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{p} \PYG{n}{PSN}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{t} \PYG{o}{\PYGZlt{}}\PYG{n}{tumor\PYGZus{}type}\PYG{o}{\PYGZgt{}}\PYG{p}{]}
\PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{m} \PYG{o}{\PYGZlt{}}\PYG{n}{medra\PYGZus{}code}\PYG{o}{\PYGZgt{}}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{O}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{o} \PYG{o}{\PYGZlt{}}\PYG{n}{results}\PYG{o}{.}\PYG{n}{txt}\PYG{o}{\PYGZgt{}}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{v}\PYG{p}{]} \PYG{p}{\PYGZob{}}\PYG{n}{patient}\PYG{p}{,}\PYG{n}{disease}\PYG{p}{\PYGZcb{}}

\PYG{n}{Positional} \PYG{n}{Arguments}\PYG{p}{:}
\PYG{n}{patient}\PYG{p}{,} \PYG{n}{disease}    \PYG{n}{Category} \PYG{n}{of} \PYG{n}{data} \PYG{n}{to} \PYG{n}{output}\PYG{o}{.} \PYG{n}{Can} \PYG{n}{either} \PYG{n}{be} \PYG{n}{patient} \PYG{o+ow}{or} \PYG{n}{disease}
\PYG{n}{level}\PYG{o}{.}

\PYG{n}{Optional} \PYG{n}{Arguments}\PYG{p}{:}
\PYG{o}{\PYGZhy{}}\PYG{n}{j}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{json}     \PYG{n}{MATCHBox} \PYG{n}{JSON} \PYG{n+nb}{file} \PYG{n}{containing} \PYG{n}{patient} \PYG{n}{data}\PYG{p}{,} \PYG{n}{usually} \PYG{k+kn}{from}
               \PYG{n+nn}{matchbox\PYGZus{}json\PYGZus{}dump.py}
\PYG{o}{\PYGZhy{}}\PYG{n}{p}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{psn}      \PYG{n}{Filter} \PYG{n}{patient} \PYG{n}{summary} \PYG{n}{to} \PYG{n}{only} \PYG{n}{these} \PYG{n}{patients}\PYG{o}{.} \PYG{n}{Can} \PYG{n}{be} \PYG{n}{a} \PYG{n}{comma}
               \PYG{n}{separated} \PYG{n+nb}{list}
\PYG{o}{\PYGZhy{}}\PYG{n}{t}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{tumor}    \PYG{n}{Retrieve} \PYG{n}{data} \PYG{k}{for} \PYG{n}{only} \PYG{n}{this} \PYG{n}{tumor} \PYG{n+nb}{type} \PYG{o+ow}{or} \PYG{n}{comma} \PYG{n}{separate} \PYG{n+nb}{list} \PYG{n}{of}
               \PYG{n}{tumors}\PYG{o}{.} \PYG{n}{Note} \PYG{n}{that} \PYG{n}{you} \PYG{n}{must} \PYG{n}{quote} \PYG{n}{tumors} \PYG{k}{with} \PYG{n}{names} \PYG{n}{containing}
               \PYG{n}{spaces}\PYG{o}{.}
\PYG{o}{\PYGZhy{}}\PYG{n}{m}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{medra}    \PYG{n}{MEDRA} \PYG{n}{Code} \PYG{o+ow}{or} \PYG{n}{comma} \PYG{n}{separated} \PYG{n+nb}{list} \PYG{n}{of} \PYG{n}{codes} \PYG{n}{to} \PYG{n}{search}\PYG{o}{.}
\PYG{o}{\PYGZhy{}}\PYG{n}{O}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{Outside}  \PYG{n}{Include} \PYG{n}{Outside} \PYG{n}{Assay} \PYG{n}{study} \PYG{n}{data} \PYG{p}{(}\PYG{n}{DEFAULT}\PYG{p}{:} \PYG{n+nb+bp}{False}\PYG{p}{)}\PYG{o}{.}
\PYG{o}{\PYGZhy{}}\PYG{n}{o}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{outfile}  \PYG{n}{Name} \PYG{n}{of} \PYG{n}{output} \PYG{n+nb}{file}\PYG{o}{.} \PYG{n}{DEFAULT}\PYG{p}{:} \PYG{n}{STDOUT}\PYG{o}{.}

\PYG{o}{\PYGZhy{}}\PYG{n}{h}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{help}     \PYG{n}{Show} \PYG{n}{this} \PYG{n}{help} \PYG{n}{message} \PYG{o+ow}{and} \PYG{n+nb}{exit}
\PYG{o}{\PYGZhy{}}\PYG{n}{v}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{version}  \PYG{n}{Show} \PYG{n}{program}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{s version number and exit}
\end{sphinxVerbatim}


\chapter{MATCHBox API Utils API Documentation}
\label{\detokenize{matchbox_api_utils::doc}}\label{\detokenize{matchbox_api_utils:matchbox-api-utils-api-documentation}}

\section{matchbox\_api\_utils.matchbox module}
\label{\detokenize{matchbox_api_utils:module-matchbox_api_utils.matchbox}}\label{\detokenize{matchbox_api_utils:matchbox-api-utils-matchbox-module}}\index{matchbox\_api\_utils.matchbox (module)}\index{Matchbox (class in matchbox\_api\_utils.matchbox)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{matchbox_api_utils:matchbox_api_utils.matchbox.Matchbox}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{matchbox\_api\_utils.matchbox.}}\sphinxbfcode{\sphinxupquote{Matchbox}}}{\emph{url}, \emph{creds}, \emph{make\_raw=None}}{}
Bases: \sphinxhref{https://docs.python.org/2.7/library/functions.html\#object}{\sphinxcode{\sphinxupquote{object}}}

MATCHBox API Connector Class

Basic connector class to make a call to the API and load the raw data. From
here we pass data, current MatchData or TreatmentArm data to appropiate
calling classes.

Used for calling to the MATCHBox API, loading data and, creating a basic
data structure. Can load a raw MATCHBox API dataset JSON file, or create
one.  Requires credentials, generally acquired from the config file generated
upon package setup.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{url}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#str}{\sphinxstyleliteralemphasis{\sphinxupquote{str}}}) \textendash{} API URL for MATCHbox. Generally only using one at the moment,
but possible to add others later.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{creds}} (\sphinxhref{https://docs.python.org/2.7/library/stdtypes.html\#dict}{\sphinxstyleliteralemphasis{\sphinxupquote{dict}}}) \textendash{} 
Username and Password credentials obtained from the config
file generated upon setup. Can also just input a dict in the form
of:

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{p}{\PYGZob{}} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{username}\PYG{l+s+s1}{\PYGZsq{}} \PYG{p}{:} \PYG{o}{\PYGZlt{}}\PYG{n}{username}\PYG{o}{\PYGZgt{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{password}\PYG{l+s+s1}{\PYGZsq{}} \PYG{p}{:} \PYG{o}{\PYGZlt{}}\PYG{n}{password}\PYG{o}{\PYGZgt{}} \PYG{p}{\PYGZcb{}}
\end{sphinxVerbatim}


\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{make\_raw}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#str}{\sphinxstyleliteralemphasis{\sphinxupquote{str}}}) \textendash{} Make a raw, unprocessed MATCHBox API JSON file. Default
filename will be raw\_mb\_obj (raw MB patient dataset) or raw\_ta\_obj (
raw treatment arm dataset) followed by a datestring. Inputting a
a string will save the file with the requested filename.

\end{itemize}

\item[{Returns}] \leavevmode
MATCHBox API dataset, used in \sphinxcode{\sphinxupquote{MatchData}} or \sphinxcode{\sphinxupquote{TreatmentArm}} classes.

\end{description}\end{quote}

\end{fulllineitems}



\section{matchbox\_api\_utils.match\_data module}
\label{\detokenize{matchbox_api_utils:module-matchbox_api_utils.match_data}}\label{\detokenize{matchbox_api_utils:matchbox-api-utils-match-data-module}}\index{matchbox\_api\_utils.match\_data (module)}\index{MatchData (class in matchbox\_api\_utils.match\_data)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{matchbox_api_utils:matchbox_api_utils.match_data.MatchData}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{matchbox\_api\_utils.match\_data.}}\sphinxbfcode{\sphinxupquote{MatchData}}}{\emph{config\_file=None}, \emph{url=None}, \emph{creds=None}, \emph{patient=None}, \emph{json\_db='sys\_default'}, \emph{load\_raw=None}, \emph{make\_raw=None}, \emph{quiet=True}}{}
Bases: \sphinxhref{https://docs.python.org/2.7/library/functions.html\#object}{\sphinxcode{\sphinxupquote{object}}}

MatchboxData class

Parsed MATCHBox Data from the API as collected from the Matchbox class above.
This class has methods to generate queries, further filtering, and heuristics
on the dataset.

Generate a MATCHBox data object that can be parsed and queried downstream
with some methods.

Can instantiate with either a config JSON file, which contains the url,
username, and password information needed to access the resource, or by
supplying the individual arguments to make the connection.  This class will
call the Matchbox class in order to make the connection and deploy the data.

Can do a live query to get data in real time, or load a MATCHBox JSON file
derived from the \sphinxcode{\sphinxupquote{matchbox\_json\_dump.py}} script that is a part of the
package.  Since data in MATCHBox is relatively static these days, it’s
preferred to use an existing JSON DB and only periodically update the DB with
a call to the aforementioned script.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{config\_file}} (\sphinxstyleliteralemphasis{\sphinxupquote{file}}) \textendash{} Custom config file to use if not using system
default.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{url}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#str}{\sphinxstyleliteralemphasis{\sphinxupquote{str}}}) \textendash{} MATCHBox API URL to use if not using a config file.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{creds}} (\sphinxhref{https://docs.python.org/2.7/library/stdtypes.html\#dict}{\sphinxstyleliteralemphasis{\sphinxupquote{dict}}}) \textendash{} 
MATCHBox credentials to use if not using a config file.
Needs to be in the form of:
\begin{quote}

\sphinxcode{\sphinxupquote{\{'username':\textless{}username\textgreater{}, 'password':\textless{}password\textgreater{}\}}}
\end{quote}


\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{patient}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#str}{\sphinxstyleliteralemphasis{\sphinxupquote{str}}}) \textendash{} Limit data to a specific PSN.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{json\_db}} (\sphinxstyleliteralemphasis{\sphinxupquote{file}}) \textendash{} MATCHbox processed JSON file containing the whole
dataset. This is usually generated from ‘matchbox\_json\_dump.py’.
The default value is \sphinxcode{\sphinxupquote{'sys\_default'}} which loads the default
package data. If you wish you get a live call, set this variable
to \sphinxcode{\sphinxupquote{None}}.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{load\_raw}} (\sphinxstyleliteralemphasis{\sphinxupquote{file}}) \textendash{} Load a raw API dataset rather than making a fresh
call to the API. This is intended for dev purpose
only and will be disabled.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{make\_raw}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#bool}{\sphinxstyleliteralemphasis{\sphinxupquote{bool}}}) \textendash{} Make a raw API JSON dataset for dev purposes only.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{quiet}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#bool}{\sphinxstyleliteralemphasis{\sphinxupquote{bool}}}) \textendash{} If True, suppress module output debug, information, etc.
messages.

\end{itemize}

\end{description}\end{quote}
\index{get\_patient\_meta() (matchbox\_api\_utils.match\_data.MatchData method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{matchbox_api_utils:matchbox_api_utils.match_data.MatchData.get_patient_meta}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_patient\_meta}}}{\emph{psn}, \emph{val=None}}{}
Return data for a patient based on a metadata field name. Sometimes we
may want to just get a quick bit of data or field for a patient record
rather than a whole analysis, and this can be a convenient way to just
check a component rather than writing a method to get each and every bit
out. If no metaval is entered, will return the whole patient dict.

\begin{sphinxadmonition}{note}{Note:}
This function is more for debugging and troubleshooting that for real
functionality.
\end{sphinxadmonition}
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{psn}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#str}{\sphinxstyleliteralemphasis{\sphinxupquote{str}}}) \textendash{} PSN of patient for which we want to receive data.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{val}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#str}{\sphinxstyleliteralemphasis{\sphinxupquote{str}}}) \textendash{} Optional metaval of data we want. If no entered, will
return the entire patient record.

\end{itemize}

\item[{Returns}] \leavevmode
Either return dict of data if a metaval entered, or whole patient
record. Returns \sphinxcode{\sphinxupquote{None}} if no data for a particular record and raises
and error if the metaval is not valid for the dataset.

\item[{Return type}] \leavevmode
\sphinxhref{https://docs.python.org/2.7/library/stdtypes.html\#dict}{dict}

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_biopsy\_summary() (matchbox\_api\_utils.match\_data.MatchData method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{matchbox_api_utils:matchbox_api_utils.match_data.MatchData.get_biopsy_summary}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_biopsy\_summary}}}{\emph{category=None}, \emph{ret\_type='counts'}}{}
Return dict of patients registered in MATCHBox with biopsy and sequencing
information.

Categories returned are total PSNs issued (including outside
assay patients), total passed biopsies, total failed biopsies (per MDACC
message), total MSNs (only counting latest MSN if more than one issued to
a biopsy due to a failure) as a method of figuring out how many NAs were
prepared, and total with sequencing data.

Can filter output based on any one criteria by leveraging the \sphinxcode{\sphinxupquote{category}}
variable
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{catetory}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#str}{\sphinxstyleliteralemphasis{\sphinxupquote{str}}}) \textendash{} 
biopsy category to return. Valid categories are:
\begin{itemize}
\item {} 
pass

\item {} 
failed\_biopsy

\item {} 
sequenced

\item {} 
outside

\item {} 
outside\_confirmation

\item {} 
progression

\item {} 
initial

\end{itemize}


\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ret\_type}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#str}{\sphinxstyleliteralemphasis{\sphinxupquote{str}}}) \textendash{} Data type to return. Valid types are “counts” and
“ids”, where counts is the total number in that category, and
ids are the BSNs for the category. Default is “counts”

\end{itemize}

\item[{Returns}] \leavevmode
Whole set of category : count or single category : count data.

\item[{Return type}] \leavevmode
\sphinxhref{https://docs.python.org/2.7/library/stdtypes.html\#dict}{dict}

\end{description}\end{quote}

\begin{sphinxadmonition}{note}{\label{matchbox_api_utils:index-0}Todo:}\begin{itemize}
\item {} 
Fix examples.

\end{itemize}
\end{sphinxadmonition}
\paragraph{Examples}

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{data}\PYG{o}{.}\PYG{n}{get\PYGZus{}biopsy\PYGZus{}summary}\PYG{p}{(}\PYG{p}{)}\PYG{p}{)}
\PYG{g+go}{\PYGZob{}u\PYGZsq{}sequenced\PYGZsq{}: 5620, u\PYGZsq{}msns\PYGZsq{}: 5620, u\PYGZsq{}progression\PYGZsq{}: 9,}
\PYG{g+go}{    u\PYGZsq{}initial\PYGZsq{}: 5563, u\PYGZsq{}patients\PYGZsq{}: 6491, u\PYGZsq{}outside\PYGZsq{}: 61,}
\PYG{g+go}{    u\PYGZsq{}no\PYGZus{}biopsy\PYGZsq{}: 465, u\PYGZsq{}failed\PYGZus{}biopsy\PYGZsq{}: 574, u\PYGZsq{}pass\PYGZsq{}: 5654,}
\PYG{g+go}{    u\PYGZsq{}outside\PYGZus{}confirmation\PYGZsq{}: 21\PYGZcb{}}
\end{sphinxVerbatim}

\end{fulllineitems}

\index{matchbox\_dump() (matchbox\_api\_utils.match\_data.MatchData method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{matchbox_api_utils:matchbox_api_utils.match_data.MatchData.matchbox_dump}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{matchbox\_dump}}}{\emph{filename=None}}{}
Dump a parsed MATCHBox dataset.

Call to the API and make a JSON file that can later be loaded in, rather
than making an API call and reprocessing. Useful for quicker look ups as
the API call can be very, very slow with such a large DB.

\begin{sphinxadmonition}{note}{Note:}
This is a different dataset than the raw dump.
\end{sphinxadmonition}
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{filename}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#str}{\sphinxstyleliteralemphasis{\sphinxupquote{str}}}) \textendash{} 
Filename to use for output. Default filename is:

\sphinxcode{\sphinxupquote{mb\_obj\_\textless{}date\_generated\textgreater{}.json}}


\item[{Returns}] \leavevmode
MATCHBox API JSON file.

\item[{Return type}] \leavevmode
json

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_psn() (matchbox\_api\_utils.match\_data.MatchData method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{matchbox_api_utils:matchbox_api_utils.match_data.MatchData.get_psn}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_psn}}}{\emph{msn=None}, \emph{bsn=None}}{}
Retrieve a patient PSN from either an input MSN or BSN.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{msn}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#str}{\sphinxstyleliteralemphasis{\sphinxupquote{str}}}) \textendash{} A MSN number to query.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{bsn}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#str}{\sphinxstyleliteralemphasis{\sphinxupquote{str}}}) \textendash{} A BSN number to query.

\end{itemize}

\item[{Returns}] \leavevmode
A PSN that maps to the MSN or BSN input.

\item[{Return type}] \leavevmode
\sphinxhref{https://docs.python.org/2.7/library/functions.html\#str}{str}

\end{description}\end{quote}
\paragraph{Examples}

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{get\PYGZus{}psn}\PYG{p}{(}\PYG{n}{bsn}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{T\PYGZhy{}17\PYGZhy{}000550}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}\PYG{p}{)}
\PYG{g+go}{PSN14420}
\end{sphinxVerbatim}

\end{fulllineitems}

\index{get\_msn() (matchbox\_api\_utils.match\_data.MatchData method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{matchbox_api_utils:matchbox_api_utils.match_data.MatchData.get_msn}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_msn}}}{\emph{psn=None}, \emph{bsn=None}}{}
Retrieve a patient MSN from either an input PSN or BSN.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{psn}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#str}{\sphinxstyleliteralemphasis{\sphinxupquote{str}}}) \textendash{} A MSN number to query.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{bsn}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#str}{\sphinxstyleliteralemphasis{\sphinxupquote{str}}}) \textendash{} A BSN number to query.

\end{itemize}

\end{description}\end{quote}

\begin{sphinxadmonition}{note}{Note:}
There can always be more than 1 MSN per patient, but can only
ever be 1 MSN per biopsy at a time.
\end{sphinxadmonition}
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
A list of MSNs that correspond with the input PSN or BSN.

\item[{Return type}] \leavevmode
list

\end{description}\end{quote}
\paragraph{Examples}

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{get\PYGZus{}msn}\PYG{p}{(}\PYG{n}{bsn}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{T\PYGZhy{}17\PYGZhy{}000550}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}\PYG{p}{)}
\PYG{g+go}{[u\PYGZsq{}MSN44180\PYGZsq{}]}
\end{sphinxVerbatim}

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{get\PYGZus{}msn}\PYG{p}{(}\PYG{n}{bsn}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{T\PYGZhy{}16\PYGZhy{}000811}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}\PYG{p}{)}
\PYG{g+go}{[u\PYGZsq{}MSN18184\PYGZsq{}]}
\end{sphinxVerbatim}

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{get\PYGZus{}msn}\PYG{p}{(}\PYG{n}{psn}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{11583}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}\PYG{p}{)}
\PYG{g+go}{[u\PYGZsq{}MSN18184\PYGZsq{}, u\PYGZsq{}MSN41897\PYGZsq{}]}
\end{sphinxVerbatim}

\end{fulllineitems}

\index{get\_bsn() (matchbox\_api\_utils.match\_data.MatchData method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{matchbox_api_utils:matchbox_api_utils.match_data.MatchData.get_bsn}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_bsn}}}{\emph{psn=None}, \emph{msn=None}}{}
Retrieve a patient BSN from either an input PSN or MSN.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{psn}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#str}{\sphinxstyleliteralemphasis{\sphinxupquote{str}}}) \textendash{} A PSN number to query.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{msn}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#str}{\sphinxstyleliteralemphasis{\sphinxupquote{str}}}) \textendash{} A MSN number to query.

\end{itemize}

\end{description}\end{quote}

\begin{sphinxadmonition}{note}{Note:}
Can have more than one BSN per PSN, but can only ever have
one BSN / MSN.
\end{sphinxadmonition}
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
A list BSNs that correspond to the PSN or MSN input.

\item[{Return type}] \leavevmode
list

\end{description}\end{quote}
\paragraph{Examples}

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{get\PYGZus{}bsn}\PYG{p}{(}\PYG{n}{psn}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{14420}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}\PYG{p}{)}
\PYG{g+go}{[u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}000550\PYGZsq{}]}
\end{sphinxVerbatim}

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{get\PYGZus{}bsn}\PYG{p}{(}\PYG{n}{psn}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{11583}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}\PYG{p}{)}
\PYG{g+go}{[u\PYGZsq{}T\PYGZhy{}16\PYGZhy{}000811\PYGZsq{}, u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}000333\PYGZsq{}]}
\end{sphinxVerbatim}

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{get\PYGZus{}bsn}\PYG{p}{(}\PYG{n}{msn}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{18184}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}\PYG{p}{)}
\PYG{g+go}{[u\PYGZsq{}T\PYGZhy{}16\PYGZhy{}000811\PYGZsq{}]}
\end{sphinxVerbatim}

\end{fulllineitems}

\index{get\_disease\_summary() (matchbox\_api\_utils.match\_data.MatchData method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{matchbox_api_utils:matchbox_api_utils.match_data.MatchData.get_disease_summary}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_disease\_summary}}}{\emph{query\_disease=None}, \emph{query\_medra=None}}{}
Return a summary of registered diseases and counts. With no args, will
return a list of all diseases and counts as a dict. One can also limit
output to a list of diseases or medra codes and get counts for those only.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{query\_disease}} (\sphinxstyleliteralemphasis{\sphinxupquote{list}}) \textendash{} List of diseases to filter on.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{query\_medra}} (\sphinxstyleliteralemphasis{\sphinxupquote{list}}) \textendash{} List of MEDRA codes to filter on.

\end{itemize}

\item[{Returns}] \leavevmode

Dictionary of disease(s) and counts in the form of:

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{p}{\PYGZob{}}\PYG{n}{medra\PYGZus{}code} \PYG{p}{:} \PYG{p}{(}\PYG{n}{ctep\PYGZus{}term}\PYG{p}{,} \PYG{n}{count}\PYG{p}{)}\PYG{p}{\PYGZcb{}}
\end{sphinxVerbatim}


\item[{Return type}] \leavevmode
\sphinxhref{https://docs.python.org/2.7/library/stdtypes.html\#dict}{dict}

\end{description}\end{quote}

\begin{sphinxadmonition}{note}{\label{matchbox_api_utils:index-1}Todo:}\begin{itemize}
\item {} 
Data is good but should try to filter outside assay data, failed
specimens, and progression biopsies (collapsed into one count) to
get a value that is closer to the accepted final count.

\end{itemize}
\end{sphinxadmonition}
\paragraph{Examples}

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{data}\PYG{o}{.}\PYG{n}{get\PYGZus{}disease\PYGZus{}summary}\PYG{p}{(}\PYG{n}{query\PYGZus{}medra}\PYG{o}{=}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{10006190}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{)}
\PYG{g+go}{\PYGZob{}\PYGZsq{}10006190\PYGZsq{}: (u\PYGZsq{}Invasive breast carcinoma\PYGZsq{}, 641)\PYGZcb{}}
\end{sphinxVerbatim}

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{data}\PYG{o}{.}\PYG{n}{get\PYGZus{}disease\PYGZus{}summary}\PYG{p}{(}\PYG{n}{query\PYGZus{}disease}\PYG{o}{=}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Invasive breast carcinoma}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{)}
\PYG{g+go}{\PYGZob{}u\PYGZsq{}10006190\PYGZsq{}: (\PYGZsq{}Invasive breast carcinoma\PYGZsq{}, 641)\PYGZcb{}}
\end{sphinxVerbatim}

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{data}\PYG{o}{.}\PYG{n}{get\PYGZus{}disease\PYGZus{}summary}\PYG{p}{(}\PYG{n}{query\PYGZus{}medra}\PYG{o}{=}\PYG{p}{[}\PYG{l+m+mi}{10006190}\PYG{p}{,}\PYG{l+m+mi}{10024193}\PYG{p}{,}\PYG{l+m+mi}{10014735}\PYG{p}{]}\PYG{p}{)}
\PYG{g+go}{\PYGZob{}\PYGZsq{}10006190\PYGZsq{}: (u\PYGZsq{}Invasive breast carcinoma\PYGZsq{}, 641),}
\PYG{g+go}{ \PYGZsq{}10014735\PYGZsq{}: (u\PYGZsq{}Endometrioid endometrial adenocarcinoma\PYGZsq{}, 124),}
\PYG{g+go}{  \PYGZsq{}10024193\PYGZsq{}: (u\PYGZsq{}Leiomyosarcoma (excluding uterine leiomyosarcoma)\PYGZsq{}, 57)\PYGZcb{}}
\end{sphinxVerbatim}

\end{fulllineitems}

\index{get\_histology() (matchbox\_api\_utils.match\_data.MatchData method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{matchbox_api_utils:matchbox_api_utils.match_data.MatchData.get_histology}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_histology}}}{\emph{psn=None}, \emph{msn=None}, \emph{bsn=None}, \emph{outside=False}, \emph{no\_disease=False}}{}
Return dict of PSN:Disease for valid biopsies.  Valid biopsies can
are defined as being only \sphinxtitleref{Passed} and can not be \sphinxtitleref{Failed}, \sphinxtitleref{No Biopsy} or
outside assay biopsies at this time.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{psn}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#str}{\sphinxstyleliteralemphasis{\sphinxupquote{str}}}) \textendash{} Optional PSN or comma separated list of PSNs on which
to filter data.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{bsn}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#str}{\sphinxstyleliteralemphasis{\sphinxupquote{str}}}) \textendash{} Optional BSN or comma separated list of BSNs on which
to filter data.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{msn}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#str}{\sphinxstyleliteralemphasis{\sphinxupquote{str}}}) \textendash{} Optional MSN or comma separated list of MSNs on which
to filter data.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{outside}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#bool}{\sphinxstyleliteralemphasis{\sphinxupquote{bool}}}) \textendash{} Also include outside assay data. Default: \sphinxcode{\sphinxupquote{False}}

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{no\_disease}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#bool}{\sphinxstyleliteralemphasis{\sphinxupquote{bool}}}) \textendash{} Return all data, even if there is no disease
indicated for the patient specimen. Default: \sphinxcode{\sphinxupquote{False}}

\end{itemize}

\item[{Returns}] \leavevmode
Dict of ID : Disease mappings. If no match for input ID,
returns \sphinxcode{\sphinxupquote{None}}.

\item[{Return type}] \leavevmode
\sphinxhref{https://docs.python.org/2.7/library/stdtypes.html\#dict}{dict}

\end{description}\end{quote}
\paragraph{Examples}

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{data}\PYG{o}{.}\PYG{n}{get\PYGZus{}histology}\PYG{p}{(}\PYG{n}{psn}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{11352}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{g+go}{\PYGZob{}\PYGZsq{}PSN11352\PYGZsq{}: u\PYGZsq{}Serous endometrial adenocarcinoma\PYGZsq{}\PYGZcb{}}
\end{sphinxVerbatim}

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{data}\PYG{o}{.}\PYG{n}{get\PYGZus{}histology}\PYG{p}{(}\PYG{n}{msn}\PYG{o}{=}\PYG{l+m+mi}{3060}\PYG{p}{)}
\PYG{g+go}{No result for id MSN3060}
\PYG{g+go}{\PYGZob{}\PYGZsq{}MSN3060\PYGZsq{}: None\PYGZcb{}}
\end{sphinxVerbatim}

\end{fulllineitems}

\index{find\_variant\_frequency() (matchbox\_api\_utils.match\_data.MatchData method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{matchbox_api_utils:matchbox_api_utils.match_data.MatchData.find_variant_frequency}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{find\_variant\_frequency}}}{\emph{query}, \emph{query\_patients=None}}{}
Find and return variant hit rates.

Based on an input query in the form of a \sphinxtitleref{variant\_type : gene} dict,
where the gene value can be a list of genes, output a list of patients
that had hits in those gene with some disease and variant information.

The return val will be unique to a patient. So, in the cases where we
have multiple biopsies from the same patient (an intitial and progression
re-biopsy for example), we will only get the union of the two sets, and
duplicate variants will not be output.  This will preven the hit rate from
getting over inflated.  Also, there is no sequence specific information
output in this version (i.e. no VAF, Coverage, etc.).  Sequence level
information for a call can be obtained from the method {\hyperref[\detokenize{matchbox_api_utils:get-variant-report}]{\sphinxcrossref{get\_variant\_report}}}
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{query}} (\sphinxhref{https://docs.python.org/2.7/library/stdtypes.html\#dict}{\sphinxstyleliteralemphasis{\sphinxupquote{dict}}}) \textendash{} 
Dictionary of variant\_type: gene mappings where:

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{o}{\PYGZhy{}}  \PYG{n}{variant} \PYG{n+nb}{type} \PYG{o+ow}{is} \PYG{n}{one} \PYG{o+ow}{or} \PYG{n}{more} \PYG{n}{of} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{snvs}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{indels}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{fusions}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
    \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{cnvs}\PYG{l+s+s1}{\PYGZsq{}}
\PYG{o}{\PYGZhy{}}  \PYG{n}{gene} \PYG{o+ow}{is} \PYG{n}{a} \PYG{n+nb}{list} \PYG{n}{of} \PYG{n}{genes} \PYG{n}{to} \PYG{n}{query}\PYG{o}{.}
\end{sphinxVerbatim}


\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{query\_patients}} (\sphinxstyleliteralemphasis{\sphinxupquote{list}}) \textendash{} List of patients for which we want to obtain
data.

\end{itemize}

\item[{Returns}] \leavevmode
Return a dict of matching data with disease and MOI information,
along with a count of the number of patients queried and the
number of biopsies queried.

\item[{Return type}] \leavevmode
\sphinxhref{https://docs.python.org/2.7/library/stdtypes.html\#dict}{dict}

\end{description}\end{quote}
\paragraph{Examples}

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{query}\PYG{o}{=}\PYG{p}{\PYGZob{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{snvs}\PYG{l+s+s1}{\PYGZsq{}} \PYG{p}{:} \PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{BRAF}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{MTOR}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{indels}\PYG{l+s+s1}{\PYGZsq{}} \PYG{p}{:} \PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{BRAF}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{MTOR}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{\PYGZcb{}}
\PYG{g+go}{find\PYGZus{}variant\PYGZus{}frequency(query)}
\end{sphinxVerbatim}

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{pprint}\PYG{p}{(}\PYG{n}{data}\PYG{o}{.}\PYG{n}{find\PYGZus{}variant\PYGZus{}frequency}\PYG{p}{(}\PYG{p}{\PYGZob{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{snvs}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{EGFR}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{indels}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{EGFR}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{\PYGZcb{}}\PYG{p}{,} \PYG{p}{[}\PYG{l+m+mi}{15232}\PYG{p}{]}\PYG{p}{)}\PYG{p}{)}
\PYG{g+go}{(\PYGZob{}\PYGZsq{}15232\PYGZsq{}: \PYGZob{}\PYGZsq{}bsns\PYGZsq{}: [u\PYGZsq{}T\PYGZhy{}17\PYGZhy{}001423\PYGZsq{}],}
\PYG{g+go}{            \PYGZsq{}disease\PYGZsq{}: u\PYGZsq{}Lung adenocarcinoma\PYGZsq{},}
\PYG{g+go}{            \PYGZsq{}mois\PYGZsq{}: [\PYGZob{}\PYGZsq{}alternative\PYGZsq{}: u\PYGZsq{}T\PYGZsq{},}
\PYG{g+go}{                      \PYGZsq{}amoi\PYGZsq{}: [u\PYGZsq{}EAY131\PYGZhy{}E(i)\PYGZsq{}, u\PYGZsq{}EAY131\PYGZhy{}A(e)\PYGZsq{}],}
\PYG{g+go}{                      \PYGZsq{}chromosome\PYGZsq{}: u\PYGZsq{}chr7\PYGZsq{},}
\PYG{g+go}{                      \PYGZsq{}confirmed\PYGZsq{}: True,}
\PYG{g+go}{                      \PYGZsq{}exon\PYGZsq{}: u\PYGZsq{}20\PYGZsq{},}
\PYG{g+go}{                      \PYGZsq{}function\PYGZsq{}: u\PYGZsq{}missense\PYGZsq{},}
\PYG{g+go}{                      \PYGZsq{}gene\PYGZsq{}: u\PYGZsq{}EGFR\PYGZsq{},}
\PYG{g+go}{                      \PYGZsq{}hgvs\PYGZsq{}: u\PYGZsq{}c.2369C\PYGZgt{}T\PYGZsq{},}
\PYG{g+go}{                      \PYGZsq{}identifier\PYGZsq{}: u\PYGZsq{}COSM6240\PYGZsq{},}
\PYG{g+go}{                      \PYGZsq{}oncominevariantclass\PYGZsq{}: u\PYGZsq{}Hotspot\PYGZsq{},}
\PYG{g+go}{                      \PYGZsq{}position\PYGZsq{}: u\PYGZsq{}55249071\PYGZsq{},}
\PYG{g+go}{                      \PYGZsq{}protein\PYGZsq{}: u\PYGZsq{}p.Thr790Met\PYGZsq{},}
\PYG{g+go}{                      \PYGZsq{}reference\PYGZsq{}: u\PYGZsq{}C\PYGZsq{},}
\PYG{g+go}{                      \PYGZsq{}transcript\PYGZsq{}: u\PYGZsq{}NM\PYGZus{}005228.3\PYGZsq{},}
\PYG{g+go}{                      \PYGZsq{}type\PYGZsq{}: u\PYGZsq{}snvs\PYGZus{}indels\PYGZsq{}\PYGZcb{},}
\PYG{g+go}{                     \PYGZob{}\PYGZsq{}alternative\PYGZsq{}: u\PYGZsq{}\PYGZhy{}\PYGZsq{},}
\PYG{g+go}{                      \PYGZsq{}amoi\PYGZsq{}: [u\PYGZsq{}EAY131\PYGZhy{}A(i)\PYGZsq{}],}
\PYG{g+go}{                      \PYGZsq{}chromosome\PYGZsq{}: u\PYGZsq{}chr7\PYGZsq{},}
\PYG{g+go}{                      \PYGZsq{}confirmed\PYGZsq{}: True,}
\PYG{g+go}{                      \PYGZsq{}exon\PYGZsq{}: u\PYGZsq{}19\PYGZsq{},}
\PYG{g+go}{                      \PYGZsq{}function\PYGZsq{}: u\PYGZsq{}nonframeshiftDeletion\PYGZsq{},}
\PYG{g+go}{                      \PYGZsq{}gene\PYGZsq{}: u\PYGZsq{}EGFR\PYGZsq{},}
\PYG{g+go}{                      \PYGZsq{}hgvs\PYGZsq{}: u\PYGZsq{}c.2240\PYGZus{}2257delTAAGAGAAGCAACATCTC\PYGZsq{},}
\PYG{g+go}{                      \PYGZsq{}identifier\PYGZsq{}: u\PYGZsq{}COSM12370\PYGZsq{},}
\PYG{g+go}{                      \PYGZsq{}oncominevariantclass\PYGZsq{}: u\PYGZsq{}Hotspot\PYGZsq{},}
\PYG{g+go}{                      \PYGZsq{}position\PYGZsq{}: u\PYGZsq{}55242470\PYGZsq{},}
\PYG{g+go}{                      \PYGZsq{}protein\PYGZsq{}: u\PYGZsq{}p.Leu747\PYGZus{}Pro753delinsSer\PYGZsq{},}
\PYG{g+go}{                      \PYGZsq{}reference\PYGZsq{}: u\PYGZsq{}TAAGAGAAGCAACATCTC\PYGZsq{},}
\PYG{g+go}{                      \PYGZsq{}transcript\PYGZsq{}: u\PYGZsq{}NM\PYGZus{}005228.3\PYGZsq{},}
\PYG{g+go}{                      \PYGZsq{}type\PYGZsq{}: u\PYGZsq{}snvs\PYGZus{}indels\PYGZsq{}\PYGZcb{}],}
\PYG{g+go}{            \PYGZsq{}msns\PYGZsq{}: [u\PYGZsq{}MSN52258\PYGZsq{}],}
\PYG{g+go}{            \PYGZsq{}psn\PYGZsq{}: u\PYGZsq{}15232\PYGZsq{}\PYGZcb{}\PYGZcb{},}
\PYG{g+go}{1)}
\end{sphinxVerbatim}

\end{fulllineitems}

\index{get\_variant\_report() (matchbox\_api\_utils.match\_data.MatchData method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{matchbox_api_utils:matchbox_api_utils.match_data.MatchData.get_variant_report}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_variant\_report}}}{\emph{psn=None}, \emph{msn=None}}{}~\phantomsection\label{\detokenize{matchbox_api_utils:get-variant-report}}
Input a PSN or MSN \sphinxstylestrong{(preferred!)} and return a list of dicts of variant
call data.

Since there can be more than one MSN per patient, one will get a more
robust result by querying on a MSN.  That is, only one variant report
per MSN can be generated and the results, then, will be clear.  In the
case of querying by PSN, a variant report for each MSN under that PSN,
assuming that the MSN is associated with a variant report, will be
returned.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{msn}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#str}{\sphinxstyleliteralemphasis{\sphinxupquote{str}}}) \textendash{} MSN for which a variant report should be returned.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{psn}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#str}{\sphinxstyleliteralemphasis{\sphinxupquote{str}}}) \textendash{} PSN for which the variant reports should be returned.

\end{itemize}

\item[{Returns}] \leavevmode

List of dicts of variant results.

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{msn}\PYG{p}{:} \PYG{p}{\PYGZob{}}
    \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{singleNucleotideVariants}\PYG{l+s+s1}{\PYGZsq{}} \PYG{p}{:} \PYG{p}{[}\PYG{p}{\PYGZob{}}\PYG{n}{var\PYGZus{}data}\PYG{p}{\PYGZcb{}}\PYG{p}{]}\PYG{p}{,}
    \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{copyNumberVariants}\PYG{l+s+s1}{\PYGZsq{}} \PYG{p}{:} \PYG{p}{[}\PYG{p}{\PYGZob{}}\PYG{n}{var\PYGZus{}data}\PYG{p}{\PYGZcb{}}\PYG{p}{,}\PYG{p}{\PYGZob{}}\PYG{n}{var\PYGZus{}data}\PYG{p}{\PYGZcb{}}\PYG{p}{]}\PYG{p}{,} \PYG{n}{etc}\PYG{o}{.}
\PYG{p}{\PYGZcb{}}
\end{sphinxVerbatim}


\item[{Return type}] \leavevmode
\sphinxhref{https://docs.python.org/2.7/library/stdtypes.html\#dict}{dict}

\end{description}\end{quote}

Examples:

\begin{sphinxadmonition}{note}{\label{matchbox_api_utils:index-2}Todo:}
Add examples here.
\end{sphinxadmonition}

\end{fulllineitems}

\index{get\_patient\_ta\_status() (matchbox\_api\_utils.match\_data.MatchData method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{matchbox_api_utils:matchbox_api_utils.match_data.MatchData.get_patient_ta_status}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_patient\_ta\_status}}}{\emph{psn=None}}{}
Input a list of PSNs and return information about the treatment arm(s)
to which they were assigned, if they were assigned to any arms. If no PSN
list is passed to the function, return results for every PSN in the study.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{psn}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#str}{\sphinxstyleliteralemphasis{\sphinxupquote{str}}}) \textendash{} PSN string to query.

\item[{Returns}] \leavevmode
Dict of Arm IDs with last status.

\item[{Return type}] \leavevmode
\sphinxhref{https://docs.python.org/2.7/library/stdtypes.html\#dict}{dict}

\end{description}\end{quote}
\paragraph{Examples}

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{data}\PYG{o}{.}\PYG{n}{get\PYGZus{}patient\PYGZus{}ta\PYGZus{}status}\PYG{p}{(}\PYG{n}{psn}\PYG{o}{=}\PYG{l+m+mi}{10837}\PYG{p}{)}
\PYG{g+go}{\PYGZob{}u\PYGZsq{}EAY131\PYGZhy{}Z1A\PYGZsq{}: u\PYGZsq{}ON\PYGZus{}TREATMENT\PYGZus{}ARM\PYGZsq{}\PYGZcb{}}
\end{sphinxVerbatim}

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{data}\PYG{o}{.}\PYG{n}{get\PYGZus{}patient\PYGZus{}ta\PYGZus{}status}\PYG{p}{(}\PYG{n}{psn}\PYG{o}{=}\PYG{l+m+mi}{11889}\PYG{p}{)}
\PYG{g+go}{\PYGZob{}u\PYGZsq{}EAY131\PYGZhy{}IX1\PYGZsq{}: u\PYGZsq{}FORMERLY\PYGZus{}ON\PYGZus{}ARM\PYGZus{}OFF\PYGZus{}TRIAL\PYGZsq{},}
\PYG{g+go}{        u\PYGZsq{}EAY131\PYGZhy{}I\PYGZsq{}: u\PYGZsq{}COMPASSIONATE\PYGZus{}CARE\PYGZsq{}\PYGZcb{}}
\end{sphinxVerbatim}

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{data}\PYG{o}{.}\PYG{n}{get\PYGZus{}patient\PYGZus{}ta\PYGZus{}status}\PYG{p}{(}\PYG{n}{psn}\PYG{o}{=}\PYG{l+m+mi}{10003}\PYG{p}{)}
\PYG{g+go}{\PYGZob{}\PYGZcb{}}
\end{sphinxVerbatim}

\end{fulllineitems}

\index{get\_patients\_by\_disease() (matchbox\_api\_utils.match\_data.MatchData method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{matchbox_api_utils:matchbox_api_utils.match_data.MatchData.get_patients_by_disease}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_patients\_by\_disease}}}{\emph{histology=None}, \emph{medra\_code=None}}{}
Input a disease and return a list of patients that were registered with
that disease type. For histology query, we can do partial matching based
on the python \sphinxcode{\sphinxupquote{in}} function. So, if one were to query \sphinxtitleref{Lung
Adenocarcinoma}, \sphinxtitleref{Lung} , \sphinxtitleref{Lung Adeno}, or \sphinxtitleref{Adeno}, all Lung Adenocarinoma
cases would be returned.

\begin{sphinxadmonition}{note}{Note:}
Simply inputting \sphinxtitleref{Lung}, would also return \sphinxtitleref{Non-small Cell Lung
Adenocarinoma}, \sphinxtitleref{Squamous Cell Lung Adenocarcinoma}, etc, and querying
\sphinxtitleref{Adeno} would return anything that had \sphinxtitleref{adeno}. So, care must be taken
with the query, and secondary filtering may be necessary. Querying
based on MEDRA codes is specific and only an exact match will return
results; \sphinxstylestrong{This is the preferred method.}
\end{sphinxadmonition}
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{histology}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#str}{\sphinxstyleliteralemphasis{\sphinxupquote{str}}}) \textendash{} One of the CTEP shotname disease codes.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{medra\_code}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#str}{\sphinxstyleliteralemphasis{\sphinxupquote{str}}}) \textendash{} A MEDRA code to query rather than histology.

\end{itemize}

\item[{Returns}] \leavevmode
Dict of Patient : Histology Mapping

\item[{Return type}] \leavevmode
\sphinxhref{https://docs.python.org/2.7/library/stdtypes.html\#dict}{dict}

\end{description}\end{quote}
\paragraph{Examples}

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{o}{\PYGZlt{}}\PYG{n}{put} \PYG{n}{example} \PYG{n}{here}\PYG{o}{\PYGZgt{}}
\end{sphinxVerbatim}

\end{fulllineitems}

\index{get\_patients\_by\_arm() (matchbox\_api\_utils.match\_data.MatchData method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{matchbox_api_utils:matchbox_api_utils.match_data.MatchData.get_patients_by_arm}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_patients\_by\_arm}}}{\emph{arm}}{}
Input an official NCI-MATCH arm identifier (e.g. \sphinxtitleref{EAY131-A}) and return
a set of patients that have ever qualified for the arm based on variant
level data.  This not only includes patients \sphinxtitleref{ON\_TREATMENT\_ARM}, but also
\sphinxtitleref{FORMERLY\_ON\_ARM\_OFF\_TRIAL} and even \sphinxtitleref{COMPASSIONATE\_CARE}.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{arm}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#str}{\sphinxstyleliteralemphasis{\sphinxupquote{str}}}) \textendash{} One of the official NCI-MATCH arm identifiers.

\item[{Returns}] \leavevmode
List of tuples of patient, arm, and arm\_status.

\item[{Return type}] \leavevmode
list

\end{description}\end{quote}
\paragraph{Examples}

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n+nb}{print}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{get\PYGZus{}patients\PYGZus{}by\PYGZus{}arm}\PYG{p}{(}\PYG{n}{arm}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{EAY131\PYGZhy{}E}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}\PYG{p}{)}
\PYG{g+go}{[}
\PYG{g+go}{    (u\PYGZsq{}11476\PYGZsq{}, \PYGZsq{}EAY131\PYGZhy{}E\PYGZsq{}, u\PYGZsq{}OFF\PYGZus{}TRIAL\PYGZus{}DECEASED\PYGZsq{}),}
\PYG{g+go}{    (u\PYGZsq{}14343\PYGZsq{}, \PYGZsq{}EAY131\PYGZhy{}E\PYGZsq{}, u\PYGZsq{}FORMERLY\PYGZus{}ON\PYGZus{}ARM\PYGZus{}OFF\PYGZus{}TRIAL\PYGZsq{}),}
\PYG{g+go}{    (u\PYGZsq{}10626\PYGZsq{}, \PYGZsq{}EAY131\PYGZhy{}E\PYGZsq{}, u\PYGZsq{}ON\PYGZus{}TREATMENT\PYGZus{}ARM\PYGZsq{}),}
\PYG{g+go}{    (u\PYGZsq{}14256\PYGZsq{}, \PYGZsq{}EAY131\PYGZhy{}E\PYGZsq{}, u\PYGZsq{}ON\PYGZus{}TREATMENT\PYGZus{}ARM\PYGZsq{}),}
\PYG{g+go}{    (u\PYGZsq{}16472\PYGZsq{}, \PYGZsq{}EAY131\PYGZhy{}E\PYGZsq{}, u\PYGZsq{}FORMERLY\PYGZus{}ON\PYGZus{}ARM\PYGZus{}OFF\PYGZus{}TRIAL\PYGZsq{})}
\PYG{g+go}{]}
\end{sphinxVerbatim}

\end{fulllineitems}

\index{get\_ihc\_results() (matchbox\_api\_utils.match\_data.MatchData method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{matchbox_api_utils:matchbox_api_utils.match_data.MatchData.get_ihc_results}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_ihc\_results}}}{\emph{psn=None}, \emph{msn=None}, \emph{bsn=None}, \emph{assays=None}}{}
Get the IHC results for a patient.

Input a PSN, MSN, or BSN and / or a set of IHC assays, and return a dict
of data.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{psn}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#str}{\sphinxstyleliteralemphasis{\sphinxupquote{str}}}) \textendash{} Query the data by PSN.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{msn}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#str}{\sphinxstyleliteralemphasis{\sphinxupquote{str}}}) \textendash{} Query the data by MSN.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{bsn}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#str}{\sphinxstyleliteralemphasis{\sphinxupquote{str}}}) \textendash{} Query the data by BSN.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{assay}} (\sphinxstyleliteralemphasis{\sphinxupquote{list}}) \textendash{} IHC assay for which we want to return results. If no
assay is passed, will return all IHC assay results.

\end{itemize}

\end{description}\end{quote}

\begin{sphinxadmonition}{note}{Note:}
Multiple results will be returned since there can be more than
one MSN / PSN.
\end{sphinxadmonition}
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
Dict of lists containing the MSN and all IHC assays available
for the specimen.

\item[{Return type}] \leavevmode
\sphinxhref{https://docs.python.org/2.7/library/stdtypes.html\#dict}{dict}

\end{description}\end{quote}
\paragraph{Examples}

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{get\PYGZus{}ihc\PYGZus{}results}\PYG{p}{(}\PYG{n}{msn}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{MSN30791}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{g+go}{\PYGZob{}\PYGZsq{}MSN30791\PYGZsq{}: \PYGZob{}\PYGZsq{}MLH1\PYGZsq{}: u\PYGZsq{}POSITIVE\PYGZsq{},}
\PYG{g+go}{              \PYGZsq{}MSH2\PYGZsq{}: u\PYGZsq{}POSITIVE\PYGZsq{},}
\PYG{g+go}{              \PYGZsq{}PTEN\PYGZsq{}: u\PYGZsq{}POSITIVE\PYGZsq{},}
\PYG{g+go}{              \PYGZsq{}RB\PYGZsq{}: u\PYGZsq{}ND\PYGZsq{}\PYGZcb{}\PYGZcb{}}
\end{sphinxVerbatim}

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{get\PYGZus{}ihc\PYGZus{}results}\PYG{p}{(}\PYG{n}{bsn}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{T\PYGZhy{}16\PYGZhy{}002222}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{assays}\PYG{o}{=}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{PTEN}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{)}
\PYG{g+go}{\PYGZob{}u\PYGZsq{}MSN30791\PYGZsq{}: \PYGZob{}\PYGZsq{}PTEN\PYGZsq{}: u\PYGZsq{}POSITIVE\PYGZsq{}\PYGZcb{}\PYGZcb{}}
\end{sphinxVerbatim}

\end{fulllineitems}


\end{fulllineitems}



\section{matchbox\_api\_utils.match\_arms module}
\label{\detokenize{matchbox_api_utils:matchbox-api-utils-match-arms-module}}\label{\detokenize{matchbox_api_utils:module-matchbox_api_utils.match_arms}}\index{matchbox\_api\_utils.match\_arms (module)}\index{TreatmentArms (class in matchbox\_api\_utils.match\_arms)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{matchbox_api_utils:matchbox_api_utils.match_arms.TreatmentArms}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{matchbox\_api\_utils.match\_arms.}}\sphinxbfcode{\sphinxupquote{TreatmentArms}}}{\emph{config\_file=None}, \emph{url=None}, \emph{creds=None}, \emph{json\_db='sys\_default'}, \emph{load\_raw=None}, \emph{make\_raw=False}}{}
Bases: \sphinxhref{https://docs.python.org/2.7/library/functions.html\#object}{\sphinxcode{\sphinxupquote{object}}}

\sphinxstylestrong{NCI-MATCH Treatment Arms and aMOIs Class}

Generate a MATCHBox treatment arms object that can be parsed and queried
downstream.

Can instantiate with either a config JSON file, which contains the url,
username, and password information needed to access the resource, or by
supplying the individual arguments to make the connection.  This class
will call the Matchbox class in order to make the connection and deploy
the data.

Can do a live query to get data in real time, or load a MATCHBox JSON file
derived from the matchbox\_json\_dump.py script that is a part of the package.
Since data in MATCHBox is relatively static these days, it’s preferred to
use an existing JSON DB and only periodically update the DB with a call
to the aforementioned script.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{config\_file}} (\sphinxstyleliteralemphasis{\sphinxupquote{file}}) \textendash{} Custom config file to use if not using system default.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{url}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#str}{\sphinxstyleliteralemphasis{\sphinxupquote{str}}}) \textendash{} MATCHBox API URL to use.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{creds}} (\sphinxhref{https://docs.python.org/2.7/library/stdtypes.html\#dict}{\sphinxstyleliteralemphasis{\sphinxupquote{dict}}}) \textendash{} 
MATCHBox credentials to use. Needs to be in the form of:

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{p}{\PYGZob{}} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{username}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:}\PYG{o}{\PYGZlt{}}\PYG{n}{username}\PYG{o}{\PYGZgt{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{password}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:}\PYG{o}{\PYGZlt{}}\PYG{n}{password}\PYG{o}{\PYGZgt{}} \PYG{p}{\PYGZcb{}}
\end{sphinxVerbatim}


\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{json\_db}} (\sphinxstyleliteralemphasis{\sphinxupquote{file}}) \textendash{} MATCHbox processed JSON file containing the whole
dataset. This is usually generated from \sphinxcode{\sphinxupquote{'matchbox\_json\_dump.py'}}.
The default value is \sphinxcode{\sphinxupquote{'sys\_default'}} which loads the default package
data. If you wish you get a live call, set this
variable to \sphinxtitleref{“None”}.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{load\_raw}} (\sphinxstyleliteralemphasis{\sphinxupquote{file}}) \textendash{} Load a raw API dataset rather than making a fresh call
to the API. This is intended for dev purpose only and will be
disabled in production.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{make\_raw}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#bool}{\sphinxstyleliteralemphasis{\sphinxupquote{bool}}}) \textendash{} Make a raw API JSON dataset for dev purposes only.

\end{itemize}

\item[{Returns}] \leavevmode
MATCH Treatment Arms object of data and methods.

\end{description}\end{quote}
\index{ta\_json\_dump() (matchbox\_api\_utils.match\_arms.TreatmentArms method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{matchbox_api_utils:matchbox_api_utils.match_arms.TreatmentArms.ta_json_dump}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{ta\_json\_dump}}}{\emph{amois\_filename=None}, \emph{ta\_filename=None}}{}
Dump the TreatmentArms data to a JSON file that can be easily loaded
downstream. We will make both the treatment arms object, as well as the
amois lookup table object.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{amois\_filename}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#str}{\sphinxstyleliteralemphasis{\sphinxupquote{str}}}) \textendash{} Name of aMOI lookup JSON file. \sphinxstylestrong{Default:}
\sphinxtitleref{amoi\_lookup\_\textless{}datestring\textgreater{}.json}

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ta\_filename}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#str}{\sphinxstyleliteralemphasis{\sphinxupquote{str}}}) \textendash{} Name of TA object JSON file \sphinxstylestrong{Default:}
\sphinxtitleref{ta\_obj\_\textless{}datestring\textgreater{}.json}

\end{itemize}

\item[{Returns}] \leavevmode
ta\_obj\_\textless{}date\textgreater{}.json
amois\_lookup\_\textless{}date\textgreater{}.json

\item[{Return type}] \leavevmode
json

\end{description}\end{quote}

\end{fulllineitems}

\index{make\_match\_arms\_db() (matchbox\_api\_utils.match\_arms.TreatmentArms method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{matchbox_api_utils:matchbox_api_utils.match_arms.TreatmentArms.make_match_arms_db}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{make\_match\_arms\_db}}}{\emph{api\_data}}{}
\sphinxstylestrong{Make a database of MATCH Treatment Arms}

Read in raw API data and create pared down JSON structure that can be
easily parsed later one.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{api\_data}} (\sphinxstyleliteralemphasis{\sphinxupquote{json}}) \textendash{} Entire raw MATCHBox API retrieved dataset.

\item[{Returns}] \leavevmode
All Arm data.

\item[{Return type}] \leavevmode
json

\end{description}\end{quote}

\end{fulllineitems}

\index{map\_amoi() (matchbox\_api\_utils.match\_arms.TreatmentArms method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{matchbox_api_utils:matchbox_api_utils.match_arms.TreatmentArms.map_amoi}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{map\_amoi}}}{\emph{variant}}{}
Input a variant dict derived from some kind and return either an aMOI id
in the form of Arm(i\textbar{}e). If variant is not an aMOI, returns \sphinxcode{\sphinxupquote{'None'}}.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{variant}} (\sphinxhref{https://docs.python.org/2.7/library/stdtypes.html\#dict}{\sphinxstyleliteralemphasis{\sphinxupquote{dict}}}) \textendash{} 
Variant dict to annotate.  Dict must have the
following keys in order to be valid:

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{o}{\PYGZhy{}} \PYG{n+nb}{type} \PYG{p}{:} \PYG{p}{[}\PYG{n}{snvs\PYGZus{}indels}\PYG{p}{,} \PYG{n}{cnvs}\PYG{p}{,} \PYG{n}{fusions}\PYG{p}{]}
\PYG{o}{\PYGZhy{}} \PYG{n}{oncomineVariantClass}
\PYG{o}{\PYGZhy{}} \PYG{n}{gene}
\PYG{o}{\PYGZhy{}} \PYG{n}{identifier} \PYG{p}{(}\PYG{n}{i}\PYG{o}{.}\PYG{n}{e}\PYG{o}{.} \PYG{n}{variant} \PYG{n}{ID} \PYG{p}{(}\PYG{n}{COSM476}\PYG{p}{)}\PYG{p}{)}
\PYG{o}{\PYGZhy{}} \PYG{n}{exon}
\PYG{o}{\PYGZhy{}} \PYG{n}{function}
\end{sphinxVerbatim}

Not all variant types will have meaningful data for these fields,
and so fields may be padded with a null char (e.g. ‘.’, ‘-‘, ‘NA’,
etc.).


\end{description}\end{quote}
\begin{description}
\item[{Returns}] \leavevmode
list:  Arm ID(s) with (i)nclusion or (e)xclusion information.

\end{description}
\paragraph{Examples}

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{variant} \PYG{o}{=} \PYG{p}{\PYGZob{}}
\PYG{g+go}{    \PYGZsq{}type\PYGZsq{} : \PYGZsq{}snvs\PYGZus{}indels\PYGZsq{},}
\PYG{g+go}{    \PYGZsq{}gene\PYGZsq{} : \PYGZsq{}BRAF\PYGZsq{},}
\PYG{g+go}{    \PYGZsq{}identifier\PYGZsq{} : \PYGZsq{}COSM476\PYGZsq{},}
\PYG{g+go}{    \PYGZsq{}exon\PYGZsq{} : \PYGZsq{}15\PYGZsq{},}
\PYG{g+go}{    \PYGZsq{}function\PYGZsq{} : \PYGZsq{}missense\PYGZsq{} ,}
\PYG{g+go}{    \PYGZsq{}oncominevariantclass\PYGZsq{} : \PYGZsq{}Hotspot\PYGZsq{}}
\PYG{g+go}{\PYGZcb{}}
\PYG{g+go}{self.map\PYGZus{}amoi(variant)}
\PYG{g+go}{[\PYGZsq{}EAY131\PYGZhy{}Y(e)\PYGZsq{}, \PYGZsq{}EAY131\PYGZhy{}P(e)\PYGZsq{}, \PYGZsq{}EAY131\PYGZhy{}N(e)\PYGZsq{}, \PYGZsq{}EAY131\PYGZhy{}H(i)\PYGZsq{}]}
\end{sphinxVerbatim}

\begin{sphinxadmonition}{note}{\label{matchbox_api_utils:index-3}Todo:}
Check the examples work.
\end{sphinxadmonition}

\end{fulllineitems}

\index{map\_drug\_arm() (matchbox\_api\_utils.match\_arms.TreatmentArms method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{matchbox_api_utils:matchbox_api_utils.match_arms.TreatmentArms.map_drug_arm}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{map\_drug\_arm}}}{\emph{armid=None}, \emph{drugname=None}}{}
Input an Arm ID or a drug name, and retun a tuple of arm, drugname, and
ID. If no arm ID or drug name is input, will return a whole table of all
arm data.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{armid}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#str}{\sphinxstyleliteralemphasis{\sphinxupquote{str}}}) \textendash{} Offcial NCI-MATCH Arm ID in the form of \sphinxtitleref{EAY131-xxx} (e.g.
‘EAY131-Z1A’).

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{drugname}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#str}{\sphinxstyleliteralemphasis{\sphinxupquote{str}}}) \textendash{} Drug name as registered in the NCI-MATCH subprotocols.
Right now, required to have the full string (e.g. ‘MLN0128(TAK-228)’
or, unfortunately, ‘Sunitinib malate (SU011248 L-malate)’), but
will work on a regex to help make this easier later on.

\end{itemize}

\item[{Returns}] \leavevmode
List of tuples or \sphinxcode{\sphinxupquote{None}}.

\item[{Return type}] \leavevmode
list

\end{description}\end{quote}
\paragraph{Examples}

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{map\PYGZus{}drug\PYGZus{}arm}\PYG{p}{(}\PYG{n}{armid}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{EAY131\PYGZhy{}Z1A}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{g+go}{(u\PYGZsq{}EAY131\PYGZhy{}Z1A\PYGZsq{}, \PYGZsq{}Binimetinib\PYGZsq{}, u\PYGZsq{}788187\PYGZsq{})}
\end{sphinxVerbatim}

\begin{sphinxadmonition}{note}{\label{matchbox_api_utils:index-4}Todo:}
Write regex for drug name mapping so that we don’t need to deal with
long, cryptic strings.
\end{sphinxadmonition}

\end{fulllineitems}

\index{get\_exclusion\_disease() (matchbox\_api\_utils.match\_arms.TreatmentArms method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{matchbox_api_utils:matchbox_api_utils.match_arms.TreatmentArms.get_exclusion_disease}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_exclusion\_disease}}}{\emph{armid}}{}
Input an arm ID and return a list of exclusionary diseases for the arm,
if there are any. Otherwise return \sphinxcode{\sphinxupquote{None}}.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{armid}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#str}{\sphinxstyleliteralemphasis{\sphinxupquote{str}}}) \textendash{} Full identifier of the arm to be queried.

\item[{Returns}] \leavevmode
List of exclusionary diseases for the arm, or \sphinxcode{\sphinxupquote{None}} if there
aren’t any.

\item[{Return type}] \leavevmode
list

\end{description}\end{quote}
\paragraph{Example}

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{get\PYGZus{}exclusion\PYGZus{}disease}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{EAY131\PYGZhy{}Z1A}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{g+go}{[u\PYGZsq{}Melanoma\PYGZsq{}, u\PYGZsq{}Colorectal Cancer\PYGZsq{}]}
\end{sphinxVerbatim}

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{get\PYGZus{}exclusion\PYGZus{}disease}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{EAY131\PYGZhy{}Y}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{g+go}{None}
\end{sphinxVerbatim}

\end{fulllineitems}

\index{get\_amois\_by\_arm() (matchbox\_api\_utils.match\_arms.TreatmentArms method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{matchbox_api_utils:matchbox_api_utils.match_arms.TreatmentArms.get_amois_by_arm}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_amois\_by\_arm}}}{\emph{arm}}{}
Input an arm identifier and return a list of aMOIs for the arm broken
down by category.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{arm}} (\sphinxhref{https://docs.python.org/2.7/library/functions.html\#str}{\sphinxstyleliteralemphasis{\sphinxupquote{str}}}) \textendash{} Arm identifier to query

\end{description}\end{quote}

Returns:

\begin{sphinxadmonition}{note}{\label{matchbox_api_utils:index-5}Todo:}
Need to finish implementing this method.  Currently only dumping a
dict.
\end{sphinxadmonition}

\end{fulllineitems}


\end{fulllineitems}



\chapter{Indices and tables}
\label{\detokenize{index:indices-and-tables}}\begin{itemize}
\item {} 
\DUrole{xref,std,std-ref}{genindex}

\item {} 
\DUrole{xref,std,std-ref}{modindex}

\item {} 
\DUrole{xref,std,std-ref}{search}

\end{itemize}


\renewcommand{\indexname}{Python Module Index}
\begin{sphinxtheindex}
\def\bigletter#1{{\Large\sffamily#1}\nopagebreak\vspace{1mm}}
\bigletter{m}
\item {\sphinxstyleindexentry{matchbox\_api\_utils.match\_arms}}\sphinxstyleindexpageref{matchbox_api_utils:\detokenize{module-matchbox_api_utils.match_arms}}
\item {\sphinxstyleindexentry{matchbox\_api\_utils.match\_data}}\sphinxstyleindexpageref{matchbox_api_utils:\detokenize{module-matchbox_api_utils.match_data}}
\item {\sphinxstyleindexentry{matchbox\_api\_utils.matchbox}}\sphinxstyleindexpageref{matchbox_api_utils:\detokenize{module-matchbox_api_utils.matchbox}}
\end{sphinxtheindex}

\renewcommand{\indexname}{Index}
\printindex
\end{document}